<a href="https://colab.research.google.com/github/joaompais-rgb/Data_Cleaning_pandas/blob/main/Cleaning_Shark_Final_1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import pandas as pd
url = "https://www.sharkattackfile.net/spreadsheets/GSAF5.xls"

In [2]:
df_shark_attacks = pd.read_excel(url)

**Por qué:** `pandas` para la tabla. No importamos nada que no usemos: el enunciado pide código limpio.


**Por qué `read_excel`:** el archivo es `.xls`, no un csv.
**Por qué guardamos el original:** `df_shark_attacks` no se toca nunca. Así podemos comparar antes y después de limpiar, y no hay que volver a descargar el archivo.

In [3]:
df_shark_attacks.shape


(7125, 23)

In [4]:
df_shark_attacks.columns


Index(['Date', 'Year', 'Type', 'Country', 'State', 'Location', 'Activity',
       'Name', 'Sex', 'Age', 'Injury', 'Fatal Y/N', 'Time', 'Species ',
       'Source', 'pdf', 'href formula', 'href', 'Case Number', 'Case Number.1',
       'original order', 'Unnamed: 21', 'Unnamed: 22'],
      dtype='object')

In [5]:
df_shark_attacks.info ()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7125 entries, 0 to 7124
Data columns (total 23 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Date            7125 non-null   object 
 1   Year            7123 non-null   float64
 2   Type            7107 non-null   object 
 3   Country         7075 non-null   object 
 4   State           6638 non-null   object 
 5   Location        6558 non-null   object 
 6   Activity        6542 non-null   object 
 7   Name            6907 non-null   object 
 8   Sex             6547 non-null   object 
 9   Age             4131 non-null   object 
 10  Injury          7089 non-null   object 
 11  Fatal Y/N       6564 non-null   object 
 12  Time            3598 non-null   object 
 13  Species         3994 non-null   object 
 14  Source          7105 non-null   object 
 15  pdf             6799 non-null   object 
 16  href formula    6794 non-null   object 
 17  href            6796 non-null   o

In [6]:
df_shark_attacks.nunique()

,0
Date,6159
Year,261
Type,14
Country,257
State,956
Location,4662
Activity,1617
Name,5828
Sex,12
Age,252


In [7]:
df_shark_attacks. head ()

,Date,Year,Type,Country,State,Location,Activity,Name,Sex,Age,...,Species,Source,pdf,href formula,href,Case Number,Case Number.1,original order,Unnamed: 21,Unnamed: 22
0,18th September,2026.0,Unprovoked,Australia,Western Australia,Sorrento Beach Perth,Swimming,Greg O'Neil,M,63,...,Great White Shark 6m (20ft),Keith Cowley: Kevin McMurray Trackingsharks.co...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1,16th September,2026.0,Unprovoked,Canada,Quebec,Off the coast of Perce Le Bilbo dive site,Diving,Unknown Male,M,?,...,Great White Shark,Keith Cowley: Kevin McMurray Trackingsharks.co...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2,14th September,2026.0,Unprovoked,Bahamas,Bimini,Bimini Island,Swimming,Unknown Austrian Tourist,F,37,...,Unknown,Kevin McMurray Trackingsharks.com: Keith Cowley,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
3,13th September,2026.0,Unprovoked,Australia,Western Australia,Geraldton,Surfing,Mel Ismail,M,50's,...,Unknown,Simon De Marchi,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
4,7th September,2026.0,Unprovoked,USA,Hawaii,Honolulu,Surfing,Wants to remain anonymous,M,24,...,Tiger Shark suspected,Keith Cowley,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


In [8]:
df_shark_attacks. tail ()

,Date,Year,Type,Country,State,Location,Activity,Name,Sex,Age,...,Species,Source,pdf,href formula,href,Case Number,Case Number.1,original order,Unnamed: 21,Unnamed: 22
7120,Before 1903,0.0,Unprovoked,AUSTRALIA,Western Australia,Roebuck Bay,Diving,male,M,NaN,...,NaN,"H. Taunton; N. Bartlett, p. 234",ND-0005-RoebuckBay.pdf,http://sharkattackfile.net/spreadsheets/pdf_di...,http://sharkattackfile.net/spreadsheets/pdf_di...,ND.0005,ND.0005,6.0,NaN,NaN
7121,Before 1903,0.0,Unprovoked,AUSTRALIA,Western Australia,NaN,Pearl diving,Ahmun,M,NaN,...,NaN,"H. Taunton; N. Bartlett, pp. 233-234",ND-0004-Ahmun.pdf,http://sharkattackfile.net/spreadsheets/pdf_di...,http://sharkattackfile.net/spreadsheets/pdf_di...,ND.0004,ND.0004,5.0,NaN,NaN
7122,1900-1905,0.0,Unprovoked,USA,North Carolina,Ocracoke Inlet,Swimming,Coast Guard personnel,M,NaN,...,NaN,"F. Schwartz, p.23; C. Creswell, GSAF",ND-0003-Ocracoke_1900-1905.pdf,http://sharkattackfile.net/spreadsheets/pdf_di...,http://sharkattackfile.net/spreadsheets/pdf_di...,ND.0003,ND.0003,4.0,NaN,NaN
7123,1883-1889,0.0,Unprovoked,PANAMA,NaN,"Panama Bay 8ºN, 79ºW",NaN,Jules Patterson,M,NaN,...,NaN,"The Sun, 10/20/1938",ND-0002-JulesPatterson.pdf,http://sharkattackfile.net/spreadsheets/pdf_di...,http://sharkattackfile.net/spreadsheets/pdf_di...,ND.0002,ND.0002,3.0,NaN,NaN
7124,1845-1853,0.0,Unprovoked,CEYLON (SRI LANKA),Eastern Province,"Below the English fort, Trincomalee",Swimming,male,M,15,...,NaN,S.W. Baker,ND-0001-Ceylon.pdf,http://sharkattackfile.net/spreadsheets/pdf_di...,http://sharkattackfile.net/spreadsheets/pdf_di...,ND.0001,ND.0001,2.0,NaN,NaN


**Por qué cada uno:**
- `shape`: el tamaño del problema. Atributo, sin paréntesis.
- `columns`: los nombres exactos, con los espacios de más (`'Species '`).
- `info()`: tipos y valores rellenos. Aquí vemos que `Year` es float y `Age` es texto.
- `nunique()`: mide el desorden. Muchos valores distintos = lo mismo escrito de varias formas.
- `head().T`: la `.T` gira la tabla; con 23 columnas se lee mucho mejor.
- `tail()`: el final del archivo, donde suele acumularse la basura.

In [9]:
len(df_shark_attacks) - df_shark_attacks.count()

,0
Date,0
Year,2
Type,18
Country,50
State,487
Location,567
Activity,583
Name,218
Sex,578
Age,2994


Primera forma, sin ordenar: total de filas menos los valores rellenos de cada columna. Las columnas salen en el orden original del archivo, que es útil para leerlas junto al
`info()` de arriba.

In [10]:
df_shark_attacks.isna().sum().sort_values(ascending=False)

,0
Unnamed: 21,7124
Unnamed: 22,7123
Time,3527
Species,3131
Age,2994
Activity,583
Sex,578
Location,567
Fatal Y/N,561
State,487


Segunda forma, la más habitual: `isna()` marca los vacíos con `True` y `sum()` los cuenta. Añadimos `sort_values(ascending=False)` para ver primero las columnas más vacías, que son las candidatas a eliminarse.

In [11]:
(len(df_shark_attacks) - df_shark_attacks.count()).sort_values(ascending=False)

,0
Unnamed: 21,7124
Unnamed: 22,7123
Time,3527
Species,3131
Age,2994
Activity,583
Sex,578
Location,567
Fatal Y/N,561
State,487


Tercera forma: la primera, pero ordenada igual que la segunda, para poder compararlas línea a línea.


**Por qué dejamos las tres:** llegan al mismo resultado por caminos distintos y sirven de
comprobación cruzada. Si los números no coincidieran, habría un error en algún sitio.
Verificar antes de decidir qué columnas se borran nos parece más seguro que fiarnos de
una sola celda.

En una versión final del código dejaríamos solo `isna().sum().sort_values(ascending=False)`,
que es la forma más legible y la que se usa habitualmente.

**Preguntas**

¿Qué columnas tienen más valores ausentes?
¿Hay filas completamente vacías?
¿Las columnas que necesitamos (Type, Country, Activity, Year, Fatal) tienen muchos valores ausentes?

**Otras Notas**

Cada fila corresponde a un ataque de tiburón registrado.
Las columnas pdf, href formula y href son nombres y enlaces de los informes en PDF de cada caso. No contienen información sobre el ataque.
original order es la numeración interna del sitio web.
Unnamed: 21 y Unnamed: 22 son columnas sin encabezado, con 1 y 2 valores en más de 6000 filas: son notas sueltas en el archivo de Excel.
Decisión: estas 8 columnas se eliminarán durante la limpieza.
Age está almacenada como texto porque contiene valores como "30s" o "teen".


**Resultados:**

* Columnas más vacías: `Unnamed: 21` (7124 vacíos) y `Unnamed: 22` (7123), prácticamente
  vacías del todo. No las usamos en el análisis.
* `Activity` tiene 583 vacíos (8,2%), `Country` tiene 50 (0,7%), `Time` tiene 3527 (49,5%).
* `Date` no tiene ningún vacío: por eso el control de filas completamente vacías dio 0.
* Las tres celdas dan los mismos números: la comprobación cruzada es correcta.

**Decisión:** las columnas casi vacías se eliminan; en las que sí usamos, los vacíos pasan
a "Unknown" en vez de borrar la fila, para no perder el ataque en el resto de los análisis.

**Aviso:** `Time` está vacía en la mitad de los casos, así que el análisis por franja
horaria se basará en unos 3600 ataques. Hay que decirlo en las conclusiones.

In [12]:
print("Filas completamente vacías:", df_shark_attacks.isna().all(axis=1).sum())

Filas completamente vacías: 0


In [13]:
print("Filas duplicadas:", df_shark_attacks.duplicated().sum())

Filas duplicadas: 0


### NOTA

**Qué hicimos:** dos comprobaciones rápidas antes de empezar a limpiar. La primera busca filas en las que **todas** las celdas están vacías. La segunda busca filas repetidas exactamente.

**Qué leemos:** 0 y 0, y los dos ceros hay que interpretarlos, porque ninguno significa lo que parece. **0 filas vacías** no significa que el archivo esté completo: significa que `Date` está rellena en las 7125 filas, y si una columna nunca está vacía, ninguna fila puede estarlo del todo. **0 duplicados** no significa que no haya ataques contados dos veces: significa que `original order` es un contador único, y basta ese número para que dos filas nunca sean idénticas. Esa columna está tapando los duplicados.

**Por qué `isna().all(axis=1)` y no `.any()`:** `axis=1` recorre cada fila de izquierda a derecha. `all()` pide que **todos** los valores estén vacíos; `any()` marcaría la fila si le faltara uno solo, y con `Time` vacía en la mitad de los casos nos quedaríamos casi sin datos.

**Por qué `duplicated().sum()`:** `duplicated()` devuelve True y False por fila, y `sum()` cuenta los True, porque True vale 1. Es la misma idea que `isna().sum()`.

**Por qué los dejamos aunque den 0:** un control que da 0 también es un resultado.

**Qué nos llevamos:** hay que repetir el control de duplicados después de quitar columnas, no solo al principio. Lo hacemos más abajo, y ahí aparecen 8.

In [14]:
len(df_shark_attacks["Age"])

7125

### NOTA

**Qué hicimos:** contamos cuántas posiciones tiene la columna.

**Qué leemos:** 7125, el total de filas. En el `info()` veíamos 4131 valores rellenos. La resta, 2994, son las edades sin registrar: el 42% del archivo.

**Por qué `len()` y no `count()`:** `len()` cuenta posiciones, rellenas y vacías. `count()` cuenta solo las rellenas. Las dos juntas miden cuánto falta.

**Qué haremos con ella:** `Age` es texto porque mezcla números con palabras ("30s", "teen"). La convertimos a número más abajo, en la limpieza de edad.

In [15]:
new_df_shark_attacks = df_shark_attacks[[
    'Date', 'Year', 'Type', 'Country', 'State', 'Location',
    'Activity', 'Sex', 'Age', 'Injury', 'Fatal Y/N', 'Time'
]].copy()
new_df_shark_attacks.head()

,Date,Year,Type,Country,State,Location,Activity,Sex,Age,Injury,Fatal Y/N,Time
0,18th September,2026.0,Unprovoked,Australia,Western Australia,Sorrento Beach Perth,Swimming,M,63,Body not recovered,Y,1015hrs
1,16th September,2026.0,Unprovoked,Canada,Quebec,Off the coast of Perce Le Bilbo dive site,Diving,M,?,Injuries to chest shoulder and back,N,1030hrs
2,14th September,2026.0,Unprovoked,Bahamas,Bimini,Bimini Island,Swimming,F,37,Serious injuries to right ankle,N,1730hrs
3,13th September,2026.0,Unprovoked,Australia,Western Australia,Geraldton,Surfing,M,50's,Foot lost in attack,N,0945hrs
4,7th September,2026.0,Unprovoked,USA,Hawaii,Honolulu,Surfing,M,24,Scrapes on inside of legs,N,1640hrs


#  LIMPIEZA

**Qué hicimos:** de las 23 columnas del archivo nos quedamos con 12, y trabajamos sobre una copia.

**Por qué quitamos cada una:**
- `pdf`, `href formula`, `href`: nombres de archivo y enlaces a los informes en PDF. No describen el ataque, solo dicen dónde está el papel.
- `Case Number`, `Case Number.1`, `original order`: numeración interna del sitio web.
- `Unnamed: 21` y `Unnamed: 22`: columnas sin encabezado con 1 y 2 valores en 7125 filas. Son notas sueltas que alguien escribió al lado de la tabla en Excel.
- `Source`: el periódico que publicó la noticia.
- `Name`: el nombre de la víctima. Dato personal que no responde a ninguna pregunta nuestra.
- `Species `: la especie de tiburón. Interesante, pero está vacía en 3131 casos y el caso de negocio no pregunta por ella.

**Por qué los dos corchetes:** el de fuera significa "dame estas columnas"; el de dentro es la lista. Con un solo corchete y varios nombres, pandas da error.

**Por qué `.copy()` va fuera de los corchetes:** así copia la tabla. Dentro, como estaba antes, copiaba la lista de nombres, y la tabla nueva podía seguir siendo una vista del original: es lo que provoca el aviso `SettingWithCopyWarning` al modificarla después.

**Qué leemos en el `shape` de abajo:** `(7125, 12)`. Las filas no cambian; solo bajan las columnas. Si el número de filas hubiera cambiado, habría un error.

In [16]:
new_df_shark_attacks.shape

(7125, 12)

**Nota**
NEGOCIO = ESCUELA DE SURF SEGURA

- Saber la mayor y menor cantidad de ataques segun:
PAIS, ESTADO, LOCALIDAD
SEASON ATTACK
HORA DEL DIA

In [17]:
before = len(new_df_shark_attacks)
new_df_shark_attacks = new_df_shark_attacks.dropna(how="all")
print("Filas borradas:", before - len(new_df_shark_attacks))

Filas borradas: 0


**Qué hicimos:** borramos las filas en las que **todas** las celdas están vacías, y contamos cuántas se fueron.

**Qué leemos:** 0 filas borradas.

**Por qué da 0 y no es un error:** `Date` está rellena en las 7125 filas. Si una columna nunca está vacía, ninguna fila puede estar vacía del todo. El control es correcto, solo que este archivo no tiene ese problema.

**Por qué `how="all"` y no `how="any"`:** `any` borra la fila si le falta **un solo** valor, y aquí nos quedaríamos casi sin datos (`Time` está vacía en la mitad de los casos). `all` borra solo las filas que no contienen nada.

**Por qué guardamos `before`:** para poder restar y enseñar cuántas filas se perdieron.

In [18]:
new_df_shark_attacks.columns

Index(['Date', 'Year', 'Type', 'Country', 'State', 'Location', 'Activity',
       'Sex', 'Age', 'Injury', 'Fatal Y/N', 'Time'],
      dtype='object')

In [19]:
new_df_shark_attacks.columns = new_df_shark_attacks.columns.str.lower().str.replace(" ", "_")


**Qué hicimos:** pasamos los nombres a minúsculas y cambiamos los espacios por guiones
bajos.

**Qué leemos:** `'Fatal Y/N'` pasa a `'fatal_y/n'`, y `'Species '`, que tenía un espacio final invisible, pasa a `'species_'`.

**Por qué:** los nombres vienen desordenados del Excel. Un espacio de más da `KeyError` .

**Por qué ahora y no más tarde:** todo lo que viene después escribe nombres de columnas. Si los cambiásemos al final, habría que corregir todas las celdas anteriores.

**Por qué `.str` delante de cada método:** `.str` es el modo texto de pandas. Cada método de texto lo necesita, por eso se encadena: `.str.lower().str.replace(...)`.

In [20]:
new_df_shark_attacks = new_df_shark_attacks.rename(columns= {'date': 'month'})
new_df_shark_attacks = new_df_shark_attacks.rename(columns= {'fatal_y/n': 'fatality'})
new_df_shark_attacks.head()

,month,year,type,country,state,location,activity,sex,age,injury,fatality,time
0,18th September,2026.0,Unprovoked,Australia,Western Australia,Sorrento Beach Perth,Swimming,M,63,Body not recovered,Y,1015hrs
1,16th September,2026.0,Unprovoked,Canada,Quebec,Off the coast of Perce Le Bilbo dive site,Diving,M,?,Injuries to chest shoulder and back,N,1030hrs
2,14th September,2026.0,Unprovoked,Bahamas,Bimini,Bimini Island,Swimming,F,37,Serious injuries to right ankle,N,1730hrs
3,13th September,2026.0,Unprovoked,Australia,Western Australia,Geraldton,Surfing,M,50's,Foot lost in attack,N,0945hrs
4,7th September,2026.0,Unprovoked,USA,Hawaii,Honolulu,Surfing,M,24,Scrapes on inside of legs,N,1640hrs


`date` pasa a llamarse `month` y `fatal_y/n` pasa a `fatality`.

**La función por dentro:** `rename(columns={viejo: nuevo})` busca cada clave del diccionario entre los nombres de las columnas y la sustituye por su valor. Los nombres que no estén en el diccionario se quedan igual, y una clave que no exista simplemente no hace nada, sin dar error. Devuelve una tabla nueva, por eso hay que reasignar con `=`.

**Por qué `rename` y no reasignar la lista entera de columnas:** reasignar la lista obliga a escribir los doce nombres y a acertar el orden.

**Por qué cambiamos `fatal_y/n`:** lleva una barra. Una barra en un nombre impide la forma corta `df.fatality` y se lee mal dentro del código.

**Qué leemos:** el `head()` confirma que los datos son exactamente los mismos y que solo cambió la cabecera.

In [21]:
new_df_shark_attacks.shape

(7125, 12)

preguntamos cuántas filas y cuántas columnas tiene la tabla.

**La función por dentro:** `shape` no es una función, es un **atributo**: por eso va sin paréntesis. Devuelve una tupla `(filas, columnas)`. Si hiciera falta usarlas por separado, `df.shape[0]` son las filas y `df.shape[1]` las columnas.

**Por qué la usamos aquí:** Después de cada paso que borra o selecciona, `shape` dice en un segundo si pasó lo que queríamos o algo más.

**Qué leemos, y qué hay que mirar exactamente:** `(7125, 12)`. Las columnas bajaron de 23 a 12, las filas siguen siendo 7125, que es la parte importante: Si aquí apareciera otro número, habría un error en las celdas anteriores.

**La costumbre que conviene coger:** un `shape` después de cada paso que borra algo. Ocupa una línea y evita descubrir en la última celda que se perdieron dos mil ataques por el camino.

In [22]:
new_df_shark_attacks.isna().sum().sort_values(ascending=False)

,0
time,3527
age,2994
activity,583
sex,578
location,567
fatality,561
state,487
country,50
injury,36
type,18


contamos los valores ausentes de cada columna y los ordenamos de más a menos.

**Las funciones por dentro:** `isna()` devuelve una tabla del mismo tamaño, llena de `True` y `False`, con `True` donde el valor está vacío. `sum()` suma por columnas: como `True` vale 1 y `False` vale 0, sumar una columna de verdadero y falso es contarla. `sort_values(ascending=False)` ordena de mayor a menor, para que las columnas más vacías aparezcan primero, que son las que obligan a decidir algo.

**Por qué la repetimos, si ya la corrimos al principio:** porque la tabla ya no es la misma. Hemos quitado once columnas, así que toca volver a preguntar con cuántos datos nos quedamos de verdad.

**Qué leemos:** `time` en cabeza con 3527 vacíos, la mitad del archivo; después `age` con 2994; y `country` con solo 50. Esos tres números van a las conclusiones, porque marcan hasta dónde se puede afirmar cada cosa.

**Qué hacemos con ellos, y por qué no se borran las filas:** un ataque sin hora registrada sigue sirviendo para el análisis por país y por estación del año. Borrarlo por un campo vacío sería tirar información buena junto con la que falta. Los vacíos pasan a "unknown" y se apartan en cada gráfico, no de la tabla.

In [23]:
new_df_shark_attacks.duplicated().sum()

np.int64(8)


contamos cuántas filas están repetidas exactamente.

**La función por dentro:** `duplicated()` recorre la tabla de arriba abajo guardando las filas que ya ha visto, y devuelve una Serie de `True` y `False`, una posición por fila: `False` la primera vez que aparece una fila, `True` si ya había salido antes. No devuelve la tabla ni la modifica, devuelve solo las marcas. Dos filas son "la misma" cuando coinciden en todas las columnas, y por eso el resultado depende de qué columnas estén dentro de la tabla. `sum()` cuenta después los `True`, porque `True` vale 1 y `False` vale 0.

**Por qué la usamos aquí:** un mismo ataque puede estar registrado dos veces. Si lo contamos dos veces, el país donde ocurrió sube en el ranking sin motivo, y el ranking de países es nuestra primera pregunta.

**Qué leemos:** 8 filas duplicadas. Antes, sobre la tabla completa, el mismo control daba 0: `original order` es un contador único, y bastaba ese número para que dos filas nunca fueran idénticas. Esa columna estaba tapando los duplicados. Un 0 no  significa "no hay duplicados"; a veces significa "hay una columna que los está tapando".

**Por qué esta función y no `drop_duplicates()` directamente:** `drop_duplicates()` borra, `duplicated()` solo marca. Con 8 casos podemos mirarlos antes de decidir, y una fila borrada no se recupera.

In [24]:
new_df_shark_attacks[new_df_shark_attacks.duplicated()]

,month,year,type,country,state,location,activity,sex,age,injury,fatality,time
4134,Reported 26-Jun-1972,1972.0,Unprovoked,AUSTRALIA,Queensland,Pancake Creek,NaN,M,NaN,FATAL,Y,NaN
4179,1971,1971.0,Unprovoked,IRAN,Khuzestan Province,"Ahvaz, on the Karun River",NaN,M,NaN,Survived,N,NaN
5047,Aug-1956,1956.0,Provoked,UNITED KINGDOM,Cornwall,The Lizard,Attempting to kill a shark with explosives,M,NaN,"FATAL, PROVOKED INCIDENT",Y,NaN
5497,Fall 1943,1943.0,Unprovoked,USA,Hawaii,"Midway Island, Northwestern Hawaiian Islands",Spearfishing,M,NaN,Calf nipped in each case,N,NaN
6279,Reported 10-Oct-1906,1906.0,Unprovoked,USA,Hawaii,NaN,Swimming,M,NaN,FATAL,Y,NaN
6521,1890,1890.0,Unprovoked,INDIA,Tamil Nadu,Tuticorin,Diving,M,NaN,No details,UNKNOWN,NaN
6775,1863,1863.0,Unprovoked,CEYLON,NaN,NaN,Fell overboard,M,NaN,FATAL,Y,NaN
7118,Before 1906,0.0,Unprovoked,AUSTRALIA,NaN,NaN,Fishing,M,NaN,FATAL,Y,NaN


En vez de contarlos, los mostramos.

**La función por dentro:** aquí no hay función nueva. Es el filtro booleano de siempre: `tabla[condición]` devuelve las filas donde la condición es `True`. Lo único distinto es de dónde sale la condición. Normalmente la escribimos nosotros, como `df[df["year"] > 2000]`; aquí la fabrica `duplicated()`, que ya devuelve exactamente eso, una Serie de `True` y `False` con una posición por fila.


**Qué leemos:** las 8 filas repetidas, una debajo de otra, con todas sus columnas.


In [25]:
new_df_shark_attacks[new_df_shark_attacks.duplicated(keep=False)]

,month,year,type,country,state,location,activity,sex,age,injury,fatality,time
4133,Reported 26-Jun-1972,1972.0,Unprovoked,AUSTRALIA,Queensland,Pancake Creek,NaN,M,NaN,FATAL,Y,NaN
4134,Reported 26-Jun-1972,1972.0,Unprovoked,AUSTRALIA,Queensland,Pancake Creek,NaN,M,NaN,FATAL,Y,NaN
4178,1971,1971.0,Unprovoked,IRAN,Khuzestan Province,"Ahvaz, on the Karun River",NaN,M,NaN,Survived,N,NaN
4179,1971,1971.0,Unprovoked,IRAN,Khuzestan Province,"Ahvaz, on the Karun River",NaN,M,NaN,Survived,N,NaN
5046,Aug-1956,1956.0,Provoked,UNITED KINGDOM,Cornwall,The Lizard,Attempting to kill a shark with explosives,M,NaN,"FATAL, PROVOKED INCIDENT",Y,NaN
5047,Aug-1956,1956.0,Provoked,UNITED KINGDOM,Cornwall,The Lizard,Attempting to kill a shark with explosives,M,NaN,"FATAL, PROVOKED INCIDENT",Y,NaN
5496,Fall 1943,1943.0,Unprovoked,USA,Hawaii,"Midway Island, Northwestern Hawaiian Islands",Spearfishing,M,NaN,Calf nipped in each case,N,NaN
5497,Fall 1943,1943.0,Unprovoked,USA,Hawaii,"Midway Island, Northwestern Hawaiian Islands",Spearfishing,M,NaN,Calf nipped in each case,N,NaN
6277,Reported 10-Oct-1906,1906.0,Unprovoked,USA,Hawaii,NaN,Swimming,M,NaN,FATAL,Y,NaN
6279,Reported 10-Oct-1906,1906.0,Unprovoked,USA,Hawaii,NaN,Swimming,M,NaN,FATAL,Y,NaN


volvemos a mirar, pero enseñando las dos filas de cada pareja.

**La función por dentro:** `keep` le dice a `duplicated()` cuál de las copias no debe marcar. Con `keep='first'`, que es lo que hace por defecto, no marca la primera y marca todas las demás. Con `keep='last'` no marca la última. Con `keep=False` no perdona ninguna: marca todas las filas que tengan alguna copia. Por eso el resultado tiene ahora el doble de filas.

**Por qué la usamos aquí:** con el comportamiento por defecto solo ves la copia, nunca el original, y sin el original no puedes compararlos.

**Qué leemos:** 16 filas, 8 parejas. Si las dos son iguales en todo, es una copia y se borra. Si una tiene un campo relleno que a la otra le falta, conviene quedarse con la más completa.

In [26]:
new_df_shark_attacks = new_df_shark_attacks.drop_duplicates()

ya los revisamos uno a uno, así que los eliminamos.

**La función por dentro:** `drop_duplicates()` usa la misma comparación que `duplicated()`, pero en vez de devolver marcas devuelve la tabla sin las filas marcadas. Tiene los mismos parámetros: `keep` decide cuál se queda y `subset=[...]` compara solo unas columnas en lugar de todas. `inplace=True` le dice que cambie la tabla original en vez de devolver una nueva, y por eso la celda no muestra nada.

**Por qué la usamos aquí:** las 8 filas repetidas inflarían el recuento de ataques del país y del estado donde ocurrieron.

**Qué leemos después:** el `duplicated().sum()` de la celda siguiente devuelve 0. Esa es la comprobación. Nunca se borra sin verificar después.

**Por qué reasignamos en vez de usar `inplace=True`:** `inplace` está desaconsejado en pandas moderno y avisa con `SettingWithCopyWarning` cuando la tabla viene de otra. Reasignar con `=` hace lo mismo y deja claro en el código que la tabla cambia.

In [27]:
new_df_shark_attacks.duplicated().sum()

np.int64(0)

In [28]:
new_df_shark_attacks.head()

,month,year,type,country,state,location,activity,sex,age,injury,fatality,time
0,18th September,2026.0,Unprovoked,Australia,Western Australia,Sorrento Beach Perth,Swimming,M,63,Body not recovered,Y,1015hrs
1,16th September,2026.0,Unprovoked,Canada,Quebec,Off the coast of Perce Le Bilbo dive site,Diving,M,?,Injuries to chest shoulder and back,N,1030hrs
2,14th September,2026.0,Unprovoked,Bahamas,Bimini,Bimini Island,Swimming,F,37,Serious injuries to right ankle,N,1730hrs
3,13th September,2026.0,Unprovoked,Australia,Western Australia,Geraldton,Surfing,M,50's,Foot lost in attack,N,0945hrs
4,7th September,2026.0,Unprovoked,USA,Hawaii,Honolulu,Surfing,M,24,Scrapes on inside of legs,N,1640hrs


- limpieza para month

In [29]:
import re

MESES = {'jan': 1, 'feb': 2, 'mar': 3, 'apr': 4, 'may': 5, 'jun': 6,
         'jul': 7, 'aug': 8, 'sep': 9, 'oct': 10, 'nov': 11, 'dec': 12}

def sacar_mes(valor):
    texto = str(valor).lower()
    encontrado = re.search(
        r'(jan|feb|mar|apr|may|jun|jul|aug|sep|oct|nov|dec)', texto)
    if encontrado:
        return MESES[encontrado.group(1)]
    return None

new_df_shark_attacks['month'] = (new_df_shark_attacks['month']
                                 .apply(sacar_mes).astype('Int64'))

new_df_shark_attacks['month'].value_counts(dropna=False).sort_index()

,count
month,
1,579
2,420
3,465
4,496
5,459
6,557
7,792
8,678
9,618


sacamos el número del mes de una columna que todavía guarda la fecha entera.

Por qué no usamos pd.to_datetime directamente: parece el camino obvio y es una trampa. El archivo mezcla fechas completas ("15-Jun-2024"), textos con prefijo ("Reported 10-Oct-2023") y valores que solo traen el año ("1903"). Un valor con solo el año se convertiría en el 1 de enero, y estaríamos inventando cientos de ataques en enero. El gráfico saldría precioso y estaría mal.

Por qué en minúsculas: str(valor).lower() hace que "Jun", "JUN" y "jun" encuentren el mismo patrón, sin tener que escribir las tres variantes.

NOTA:la cuenta de ataques por mes y cuántos quedan sin mes. Ese número va a las conclusiones, porque el análisis por estación se basa solo en los que sí lo tienen.

In [30]:
HEMISFERIO_SUR = ['AUSTRALIA', 'SOUTH AFRICA', 'NEW ZEALAND', 'BRAZIL',
                  'ARGENTINA', 'CHILE', 'PERU', 'URUGUAY', 'FIJI',
                  'INDONESIA', 'PAPUA NEW GUINEA', 'MOZAMBIQUE',
                  'NAMIBIA', 'REUNION', 'REUNION ISLAND', 'MAURITIUS',
                  'NEW CALEDONIA', 'FRENCH POLYNESIA', 'MADAGASCAR',
                  'VANUATU', 'TONGA', 'SAMOA', 'AMERICAN SAMOA',
                  'SOLOMON ISLANDS']

OPUESTO = {'invierno': 'verano', 'verano': 'invierno',
           'primavera': 'otoño', 'otoño': 'primavera'}

def sacar_estacion(fila):
    mes = fila['month']
    if pd.isna(mes):
        return 'unknown'
    if mes in (12, 1, 2):
        estacion = 'invierno'
    elif mes in (3, 4, 5):
        estacion = 'primavera'
    elif mes in (6, 7, 8):
        estacion = 'verano'
    else:
        estacion = 'otoño'
    # country todavía no está limpia: "Australia" y "AUSTRALIA " tienen que contar igual
    pais = str(fila['country']).strip().upper().replace('?', '').strip()
    if pais in HEMISFERIO_SUR:
        estacion = OPUESTO[estacion]
    return estacion

new_df_shark_attacks['season'] = new_df_shark_attacks.apply(sacar_estacion, axis=1)

# comprobación: el mismo país escrito de dos maneras
print(new_df_shark_attacks['country'].str.strip().value_counts().filter(regex='(?i)australia'))
new_df_shark_attacks['season'].value_counts(dropna=False)

country
AUSTRALIA    1480
Australia      56
Name: count, dtype: int64


,count
season,
verano,2684
otoño,1686
primavera,1313
invierno,838
unknown,596


Convertimos el mes en estación, y la invertimos en los países del hemisferio sur.

**Por qué el hemisferio es imprescindible aquí:** junio es verano en Estados Unidos e invierno en Australia. Como Australia y Sudáfrica están entre los países con más ataques, sin corregirlo la mitad de nuestros datos quedaría con la estación cambiada, y la conclusión sobre cuándo reforzar la vigilancia sería exactamente la contraria de la verdadera.

**La función por dentro, y por qué `axis=1`:** hasta ahora `apply` trabajaba sobre una columna y recibía un valor. Aquí lo aplicamos sobre la tabla entera con `axis=1`, y entonces recibe una fila completa, de la que podemos leer el mes y el país a la vez. Sin `axis=1` recorrería columnas en vez de filas.

**Por qué ponemos el país en mayúsculas dentro de la función:** esta celda va antes de limpiar `country`, y el archivo escribe el mismo país de dos maneras: los registros antiguos en mayúsculas ("AUSTRALIA") y los recientes no ("Australia"). La lista está en mayúsculas, así que sin el `upper()` los ataques recientes de Australia no se invertían: la primera fila del archivo, Perth en septiembre, salía como otoño cuando allí es primavera. La comprobación de la salida enseña cuántas filas hay escritas de cada forma.

**Por qué la lista es más larga que antes:** faltaban Nueva Caledonia y la Polinesia Francesa, que están entre los países con más ataques y en el hemisferio sur, y otras islas del Pacífico sur y del Índico.

**Qué leemos:** la comprobación encuentra 1480 filas de Australia escritas «AUSTRALIA» y 56 escritas «Australia»: esas 56 son las que la versión anterior no invertía. En la tabla completa, todavía sin filtrar por años, salen 2684 ataques en verano, 1686 en otoño, 1313 en primavera y 838 en invierno, y 596 sin mes, que quedan como `unknown`.

NOTA: la lista de países del hemisferio sur la escribimos nosotros y es una simplificación. Países cruzados por el ecuador, como Indonesia, no tienen una estación única.

- limpieza para year

### El año: número, y solo años completos

**Qué hacemos:** convertimos `year` en número entero y nos quedamos con los años de nuestros tres bloques, de 1951 a 2025.

**Las funciones por dentro:** `pd.to_numeric(columna, errors='coerce')` intenta convertir cada valor en número y, gracias a `errors='coerce'`, pone vacío en lo que no pueda convertir en vez de parar con un error. `astype('Int64')`, con I mayúscula, es el entero de pandas que acepta vacíos; el `int` normal daría error mientras quede uno. `between(ANIO_MIN, ANIO_MAX)` devuelve `True` para los años dentro del intervalo, con los dos extremos incluidos.

**Por qué número y no texto:** como texto, `max()` compara letras y no cantidades, y un filtro como `>= 1951` da error. La evolución por años necesita número.

**Por qué el máximo es 2025 y no el de los datos:** el archivo llega a septiembre de 2026, y un año sin octubre, noviembre y diciembre no se puede comparar con uno completo. Además el archivo se actualiza cada semana: si calculáramos el máximo con `max()`, en enero saldría 2027 y todos los números de la presentación cambiarían solos. Un número fijo hace el notebook reproducible.

**Por qué el mínimo es 1951:** no lo elegimos, sale del máximo. Tres bloques de 25 años que terminan en 2025 empiezan en 2025 − 74 = 1951.

**Qué leemos en la salida:** cuántas filas había y cuántas quedan, y el año mínimo y el máximo. Tienen que salir 1951 y 2025: ese print es la prueba de que el filtro hizo lo que decimos.

In [31]:
ANIO_MAX = 2025                # último año completo: 2026 va por septiembre
ANIO_MIN = ANIO_MAX - 74       # 1951: tres bloques exactos de 25 años

new_df_shark_attacks['year'] = pd.to_numeric(
    new_df_shark_attacks['year'], errors='coerce').astype('Int64')

antes = len(new_df_shark_attacks)
new_df_shark_attacks = new_df_shark_attacks[
    new_df_shark_attacks['year'].between(ANIO_MIN, ANIO_MAX)]

print('Antes:', antes, '| Después:', len(new_df_shark_attacks))
print('Año mínimo:', new_df_shark_attacks['year'].min(),
      '| Año máximo:', new_df_shark_attacks['year'].max())

Antes: 7117 | Después: 5199
Año mínimo: 1951 | Año máximo: 2025


In [32]:
def agrupar_años(year):
    if pd.isna(year):
        return None
    if year > ANIO_MAX or year < ANIO_MIN:
        return None                                  # fuera de los tres bloques
    if year >= ANIO_MAX - 24:
        return f'{ANIO_MAX - 24}-{ANIO_MAX}'         # 2001-2025
    elif year >= ANIO_MAX - 49:
        return f'{ANIO_MAX - 49}-{ANIO_MAX - 25}'    # 1976-2000
    else:
        return f'{ANIO_MAX - 74}-{ANIO_MAX - 50}'    # 1951-1975

In [33]:
new_df_shark_attacks['bloque_año'] = new_df_shark_attacks['year'].apply(agrupar_años)
BLOQUE_RECIENTE = f'{ANIO_MAX - 24}-{ANIO_MAX}'

new_df_shark_attacks['bloque_año'].value_counts(dropna=False).sort_index()

,count
bloque_año,
1951-1975,1262
1976-2000,1223
2001-2025,2714


**Qué hacemos:** partimos los años en tres bloques de 25 y guardamos el resultado en una columna nueva, `bloque_año`.

**La función por dentro:** `agrupar_años` recibe un año y devuelve el nombre de su bloque. Las preguntas van de la más reciente a la más antigua: el primer `if` que se cumple gana, así que un año de 2010 entra en el primero y ya no llega a los demás. Los bordes no están escritos a mano: salen de `ANIO_MAX`, y si algún día se cambia el año máximo, los tres bloques se mueven solos. `apply` aplica la función a cada año de la columna, uno a uno.

**Por qué bloques iguales:** para comparar tramos de tiempo, los tramos tienen que medir lo mismo. Un bloque de 25 años al lado de uno de 2 no dice nada.

**Por qué `BLOQUE_RECIENTE`:** es el nombre del bloque 2001-2025 guardado en una variable. El análisis lo usa en varias celdas, y así nadie tiene que escribir el texto a mano ni equivocarse con un guion.

**Qué leemos:** 1262 ataques en 1951-1975, 1223 en 1976-2000 y 2714 en 2001-2025, y ningún `None`: el filtro de arriba ya dejó solo años de los tres bloques. El último bloque tiene más del doble que cada uno de los anteriores, y eso es sobre todo registro.

NOTA: el recuento sube muchísimo hacia el presente. Eso **no** significa que haya más tiburones: significa que hoy se registra mejor. Los recuentos entre bloques no son comparables; la mortalidad, que es un porcentaje dentro de cada bloque, sí.

- limpieza para type

In [34]:
new_df_shark_attacks['type'].unique()

array(['Unprovoked', 'Provoked', 'Questionable', 'unprovoked',
       ' Provoked', 'Watercraft', 'Sea Disaster', nan, '?', 'Unconfirmed',
       'Unverified', 'Invalid', 'Under investigation', 'Boat'],
      dtype=object)

pedimos la lista de valores distintos de la columna.

**La función por dentro:** `unique()` recorre la columna una vez y devuelve un array con cada valor la primera vez que aparece, en el orden en que aparecen, no alfabético y sin contar cuántos hay. Devuelve un array de NumPy, no una Serie, y por eso el resultado sale entre corchetes y con `dtype=object` al final.

**Por qué la usamos aquí:** antes de escribir cualquier limpieza hay que saber qué hay dentro. `unique()` es la función que enseña el desorden, y es la que dicta el diccionario que escribiremos después.

**Qué leemos:** 15 valores para lo que en realidad son 6 o 7 categorías, con tres problemas distintos a la vista. Primero, mayúsculas y espacios: `Unprovoked`, `UNprovoked` y `unprovoked` son la misma categoría escrita de tres maneras, y `' Provoked'` lleva un espacio delante. Segundo, categorías que significan lo mismo: `Questionable`, `Unconfirmed`, `Unverified`, `Under investigation` y `?`. Tercero, `Boat` y `Watercraft`, que son la misma cosa.



In [35]:
new_df_shark_attacks['type'].value_counts(dropna=False)

,count
type,
Unprovoked,3924
Provoked,493
Invalid,378
Watercraft,252
Sea Disaster,98
Questionable,26
NaN,14
Boat,7
Provoked,2


contamos los casos de cada valor.

**La función por dentro:** `value_counts()` agrupa los valores iguales, los cuenta y devuelve una Serie ya ordenada de mayor a menor, en la que el índice son los valores y los números son las frecuencias. Tiene tres parámetros que nos sirven: `dropna=False` incluye los vacíos como una categoría más, `normalize=True` devuelve proporciones en vez de conteos y `ascending=True` invierte el orden.

**Por qué la usamos aquí, después de `unique()` y no en su lugar:** `unique()` dice qué categorías hay; `value_counts()` dice cuánto pesa cada una. Juntas deciden qué merece la pena limpiar: una categoría con 4000 casos y otra con 2 no se tratan igual, y sin los números no se sabe cuál es cuál.

**Por qué `dropna=False`:** por defecto los vacíos no aparecen, y desaparecer no es lo mismo que no existir. Una columna puede parecer limpia solo porque la mitad está vacía.

**Qué leemos:** el grueso está en `Unprovoked`, y las variantes mal escritas tienen pocos casos cada una.

In [36]:
new_df_shark_attacks['type'] = new_df_shark_attacks['type'].astype(str).str.strip().str.lower()

 dejamos toda la columna escrita de la misma manera, antes de tocar el contenido.

**Las funciones por dentro:** `astype(str)` convierte cada valor a texto, y hace falta porque los métodos de texto solo funcionan sobre texto y esta columna tiene vacíos, que son `float`. `.str` es el accesorio de texto de pandas: aplica a cada valor de la columna el método que le pongas detrás, y por eso hay que repetirlo, `.str.strip().str.lower()`. `strip()` quita los espacios del principio y del final, no los de en medio. `lower()` pasa todas las letras a minúsculas.

**Por qué antes del diccionario y no después:** si limpiásemos después, el diccionario tendría que listar todas las variantes de cada palabra (`Boat`, `boat`, `BOAT`, `boat `). Normalizando primero, basta una entrada por categoría.

**Qué leemos:** de 15 valores distintos bajamos a 10 sin haber escrito todavía ni un diccionario. Estas dos funciones hacen la mitad del trabajo.

**El efecto secundario que hay que conocer:** `astype(str)` convierte los vacíos en el texto `'nan'`. No desaparecen, se disfrazan de palabra. Por eso el diccionario de la celda siguiente lleva la entrada `'nan': 'unknown'`.

In [37]:
unificar = {'boat': 'watercraft', 'questionable': 'invalid', 'unconfirmed': 'invalid', 'unverified': 'invalid', 'under investigation': 'invalid', '?': 'invalid', 'nan': 'unknown'}

In [38]:
new_df_shark_attacks['type'] = new_df_shark_attacks['type'].replace(unificar)


 escribimos un diccionario `unificar` y lo aplicamos con `replace`.

**La función por dentro:** `replace(diccionario)` recorre la columna y, cada vez que encuentra una clave del diccionario, la cambia por su valor. Lo que no está en el diccionario se queda exactamente como estaba. Devuelve una columna nueva, y por eso hay que reasignarla con `=`; si no, el cambio se pierde.

**Por qué un diccionario y no varias llamadas seguidas:** un diccionario hace todos los cambios en una pasada y deja las decisiones escritas en un solo sitio, donde se pueden leer y discutir. Cinco `replace` encadenados hacen lo mismo pero no se pueden enseñar en una diapositiva.

**Las decisiones, una a una:** `boat` pasa a `watercraft`, porque es la misma cosa escrita de dos maneras. `questionable`, `unconfirmed`, `unverified`, `under investigation` y `?` pasan a `invalid`, porque todas significan "el caso no está confirmado" y agrupadas se excluyen de una vez. Y `nan` pasa a `unknown`: son los vacíos que `astype(str)` había convertido en texto.

**Por qué `replace` y no `map`:** las dos reciben un diccionario, pero se comportan al revés con lo que no está en él. `replace` lo deja igual; `map` lo deja vacío. Aquí queremos conservar `unprovoked` y `provoked` sin escribirlos, así que `replace`.

In [39]:
unificar_final = {'unknown': 'invalid','watercraft': 'watercraft & sea disaster','sea disaster': 'watercraft & sea disaster'}
new_df_shark_attacks['type'] = new_df_shark_attacks['type'].replace(unificar_final)

 un segundo diccionario que junta lo que todavía estaba suelto.

**La función por dentro:** el mismo `replace` de la celda anterior. Lo único que cambia es el diccionario, y que ahora las claves son categorías ya limpias en lugar de erratas.

**Las dos decisiones:** `watercraft` y `sea disaster` pasan a `watercraft & sea disaster`, porque los dos son accidentes de embarcación; separados no aportan nada a nuestro caso, y juntos quedan como un grupo único que se excluye de una vez. Y `unknown` pasa a `invalid`: los casos sin tipo se meten con los no confirmados.

**El porqué de esa segunda decisión, que hay que poder defender:** `unknown` y `invalid` no significan lo mismo. `unknown` es "no sabemos qué pasó"; `invalid` es "el caso no es válido". Los juntamos porque para nuestra pregunta el efecto es idéntico: ninguno de los dos entra en el análisis, que se hace solo sobre `unprovoked`. Si la pregunta fuera sobre la calidad del registro, habría que mantenerlos separados.


In [40]:
new_df_shark_attacks['type'].value_counts(dropna=False)

,count
type,
unprovoked,3925
provoked,495
invalid,422
watercraft & sea disaster,357


volvemos a contar las categorías, ahora ya limpias.

**La función por dentro:** el mismo `value_counts(dropna=False)` de antes. No hay nada nuevo: el valor de esta celda no está en la función, está en compararla con la de antes de limpiar.

**Qué leemos** el total de casos tiene que ser el mismo que antes. Si hubiera bajado, algo se habría convertido en vacío por el camino y estaríamos perdiendo ataques sin darnos cuenta. Agrupar categorías cambia el reparto, nunca el total.

**De 15 valores a 4 categorías:** `unprovoked`, `provoked`, `invalid` y `watercraft & sea disaster`. Es la diapositiva más fácil de la presentación: la misma columna antes y después.

- limpieza para country

In [41]:
country = new_df_shark_attacks['country'].astype(str).str.strip().str.upper()
country = country.str.replace('?', '', regex=False).str.strip()

In [42]:
unificar_paises = {
    'UNITED STATES': 'USA',
    'UNITED STATES OF AMERICA': 'USA',
    'ENGLAND': 'UNITED KINGDOM',
    'SCOTLAND': 'UNITED KINGDOM',
    'WALES': 'UNITED KINGDOM',
    'COLUMBIA': 'COLOMBIA',
    'REUNION ISLAND': 'REUNION',
    'MALDIVE ISLANDS': 'MALDIVES',
    'NAN': 'UNKNOWN',
}

In [43]:
new_df_shark_attacks['country'] = country.replace(unificar_paises)
new_df_shark_attacks['country'].value_counts(dropna=False).head(15)

,count
country,
USA,2204
AUSTRALIA,961
SOUTH AFRICA,484
BAHAMAS,126
PAPUA NEW GUINEA,113
BRAZIL,113
NEW ZEALAND,99
MEXICO,80
REUNION,60


**Por qué:** `country` es la primera de las cuatro preguntas del caso de negocio. Sin ella no hay ranking de países, y sin ranking no hay respuesta a "dónde no abrir".

**Los problemas que tiene:** los mismos que `state`, y uno más. El mismo país aparece escrito de varias maneras (`USA`, `United States`, `United States of America`, `REUNION` y `REUNION ISLAND`), algunos llevan un `?` de duda al final, y hay 50 vacíos.

**Las funciones, y por qué en este orden:**
1. `strip()` y `upper()` primero. Usamos `upper()` y no `lower()` porque el archivo ya viene casi todo en mayúsculas: así se cambian menos valores y hay menos ocasiones de romper algo.
2. `replace('?', '', regex=False)` quita el signo de duda. `regex=False` le dice que el `?` es texto literal; sin eso, en expresiones regulares `?` significa "el carácter anterior, cero o una vez". Después otro `strip()`, porque al quitar el `?` puede quedar un espacio suelto.
3. El diccionario al final, cuando el texto ya está uniforme. Si lo pusiéramos primero habría que escribir cada variante en todas sus mayúsculas posibles.

**Las dos entradas nuevas:** `REUNION ISLAND` y `MALDIVE ISLANDS` son el mismo sitio que `REUNION` y `MALDIVES`. Lo descubrimos al verificar la lista de destinos: separados, cada nombre se queda con una parte de los ataques.

**Qué leemos:** Estados Unidos (2204), Australia (961) y Sudáfrica (484) concentran más de dos tercios de los ataques de la tabla. Muy por detrás vienen Bahamas (126), Papúa Nueva Guinea (113) y Brasil (113).

**Por qué los vacíos pasan a `UNKNOWN` y no se borra la fila:** son solo 50, pero un ataque sin país sigue sirviendo para contar por estación del año o por franja horaria. Los `UNKNOWN` se apartan en cada ranking, no de la tabla.

- limpieza para state

In [44]:
new_df_shark_attacks['state'].value_counts(dropna=False)

,count
state,
Florida,1106
New South Wales,310
California,303
Hawaii,282
NaN,248
...,...
Camaguey Province,1
Santiago de Cuba Province,1
Santa Isabel Province,1


contamos los casos de cada estado.

**La función por dentro:** `value_counts(dropna=False)`, otra vez. Lo que nos interesa aquí no es el primer puesto, sino cuántas filas tiene el resultado: eso es el número de valores distintos de la columna. `nunique()` daría ese número directamente, pero `value_counts()` da además el reparto, que es lo que hace falta para decidir dónde cortar.

**Qué leemos:** el archivo original tiene 956 estados distintos para 7125 ataques (lo dice el `nunique()` del principio), un valor nuevo cada siete filas. Con los años ya filtrados siguen siendo cientos: la salida de arriba lo dice en `Length`. Eso no es una columna categórica, es texto libre con erratas, mayúsculas mezcladas y nombres escritos de varias maneras.

Como hay cientos de estados distintos, nos quedamos con los 10 con más ataques y el resto va a una misma categoría, 'otros'.

In [45]:
new_df_shark_attacks['state'] = new_df_shark_attacks['state'].str.split(',').str[0]

In [46]:
new_df_shark_attacks['state'] = (new_df_shark_attacks['state'].astype(str).str.replace(r'[^\w\s]', '', regex=True).str.strip().str.replace(r'\s+', ' ', regex=True).str.lower())

normalizamos los nombres

**Qué hacemos:** `astype(str)`, `strip()` y `lower()`, las mismas tres funciones que en `type`.

**Por qué antes de calcular el top 10 y no después:** si no, "Florida", "florida" y "FLORIDA " serían tres estados distintos, cada uno con su trozo del recuento.

**El efecto secundario, otra vez:** `astype(str)` convierte los estados vacíos en el texto `'nan'`. Como ahora es una palabra, `value_counts()` la cuenta, y con cientos de casos entraría en el top 10 como si fuera un estado. Conviene ponerle nombre antes de calcular el ranking, con `replace('nan', 'unknown')`.

**Para la presentación:** `lower()` deja "new south wales" en minúsculas. En los gráficos queda mejor `.str.title()`, que pone en mayúscula la primera letra de cada palabra y devuelve "New South Wales".

In [47]:
top_estados = new_df_shark_attacks['state'].value_counts().index[:10]
top_estados

Index(['florida', 'new south wales', 'california', 'hawaii', 'nan',
       'queensland', 'western australia', 'western cape province',
       'kwazulunatal', 'eastern cape province'],
      dtype='object', name='state')

guardamos los 10 estados con más ataques.

**La función por dentro:** `value_counts()` devuelve una Serie ordenada de mayor a menor en la que los nombres de los estados son el índice y los conteos son los valores. De ahí salen las dos piezas: `.index` son los nombres (`.values` serían los números) y `[:10]` es un corte de Python, que se queda con los diez primeros. El resultado no es una tabla ni una columna, es una lista de diez nombres guardada en una variable. Esta celda no cambia nada del DataFrame.

**Por qué la usamos aquí:** porque la agrupación necesita saber primero quiénes son los diez grandes. Separar el cálculo de la lista y su aplicación deja ver la decisión antes de ejecutarla.

**Qué falta para que la agrupación se aplique:** marcar como "otros" todo lo que no esté en la lista, con `isin` y `where`. `isin(lista)` devuelve `True` donde el valor está en la lista y `False` donde no. `where(condición, otro_valor)` conserva el valor original donde la condición es `True` y pone `otro_valor` donde es `False`: va al revés de lo que parece, porque where conserva, no selecciona.

NOTA: agrupar en "otros" responde muy bien a dónde NO abrir, porque los estados peligrosos son justamente los que quedan con nombre. Pero borra la respuesta a dónde SÍ abrir: los sitios con pocos ataques son exactamente los que desaparecen dentro de "otros".Guardar el resultado en `state_group` en vez de encima de `state` deja `state_group` para los gráficos y `state` entera para buscar destinos concretos. Es la misma pareja que `Activity` y `Activity_Group`.

**Qué leemos:** `'nan'` aparece en el quinto puesto, como si fuera un estado: son los estados vacíos que `astype(str)` convirtió en texto. Esta lista es la primera versión y la dejamos para enseñar el problema; la definitiva, sin `'nan'`, se calcula más abajo.

In [48]:
new_df_shark_attacks['state'] = new_df_shark_attacks['state'].replace('nan', 'unknown')


In [49]:
top_estados = (new_df_shark_attacks['state']
               .value_counts()
               .drop('unknown', errors='ignore')
               .index[:10])

In [50]:
new_df_shark_attacks['state_group'] = new_df_shark_attacks['state'].where(
    new_df_shark_attacks['state'].isin(top_estados), 'otros')

In [51]:
new_df_shark_attacks['state_group'].value_counts()

,count
state_group,
otros,2171
florida,1107
new south wales,312
california,303
hawaii,282
queensland,212
western australia,201
western cape province,167
kwazulunatal,159


ponemos nombre a los vacíos, recalculamos el top 10 sin ellos y marcamos como
`otros` todo lo que no esté en la lista.

**Las funciones por dentro:** `isin(lista)` devuelve `True` donde el valor está en la lista y
`False` donde no. `where(condición, otro_valor)` conserva el valor original donde la condición es
`True` y pone `otro_valor` donde es `False`: va al revés de lo que parece, porque **where
conserva**, no selecciona. Y `drop('unknown', errors='ignore')` quita esa fila del recuento;
`errors='ignore'` evita que dé error si no existe.

NOTA: once categorías, las diez grandes más `otros`. Si `otros` se queda con más de la
mitad de los ataques, conviene subir a un top 15.

- limpieza para location

In [52]:
new_df_shark_attacks['location'].value_counts(dropna=False)

,count
location,
NaN,294
"New Smyrna Beach, Volusia County",191
"Daytona Beach, Volusia County",30
"Ponce Inlet, Volusia County",28
"Myrtle Beach, Horry County",22
...,...
Hanalei Bay,1
Bittern,1
Tahunanui Beach,1


In [53]:
new_df_shark_attacks['location'] = new_df_shark_attacks['location'].str.split(',').str[0]

In [54]:
new_df_shark_attacks['location'] = (new_df_shark_attacks['location']
                                 .astype(str)
                                 .str.replace(r'[^\w\s]', '', regex=True)
                                 .str.strip()
                                 .str.replace(r'\s+', ' ', regex=True)
                                 .str.lower())

dos pasos. Primero nos quedamos con lo que hay antes de la primera coma, y después quitamos la puntuación y los espacios de más.

**Las funciones por dentro:** `str.split(',')` parte cada texto en trozos cada vez que encuentra una coma y devuelve una lista; `.str[0]` coge el primer trozo de cada lista. Después, `str.replace(r'[^\w\s]', '', regex=True)` borra todo lo que no sea letra, número o espacio, porque `[^...]` significa "cualquier cosa que no sea esto". Y `str.replace(r'\s+', ' ', regex=True)` convierte cualquier grupo de espacios seguidos en uno solo, que es lo que arregla los dobles espacios que quedan al borrar signos.

**Qué leemos antes de limpiar:** la playa con más registros es New Smyrna Beach, en el condado de Volusia (Florida), y detrás vienen otras dos del mismo condado, Daytona Beach y Ponce Inlet.

NOTA: quedan los vacíos, que `astype(str)` convierte en la palabra "nan", y las variantes de nombre que no se diferencian por la coma. Con miles de valores distintos no vamos a unificarlas a mano: usaremos la columna para contar playas distintas y mirar los primeros puestos, no como categoría cerrada.

- limpieza para activity

In [55]:
def clasificar_actividad(texto):
    if pd.isna(texto):
        return 'unknown'
    texto = str(texto).lower()
    if 'surf' in texto or 'boogie' in texto or 'body board' in texto \
            or 'bodyboard' in texto or 'paddle' in texto or 'kite' in texto:
        return 'surfing'
    elif 'swim' in texto or 'bath' in texto or 'wading' in texto \
            or 'standing' in texto or 'floating' in texto:
        return 'swimming'
    elif 'fish' in texto:
        return 'fishing'
    elif 'div' in texto or 'snorkel' in texto or 'scuba' in texto:
        return 'diving'
    else:
        return 'other'

In [56]:
new_df_shark_attacks['activity_group'] = (
    new_df_shark_attacks['activity'].apply(clasificar_actividad))

In [57]:
new_df_shark_attacks['activity_group'].value_counts(dropna=False)

,count
activity_group,
surfing,1593
swimming,1151
fishing,1022
other,555
diving,529
unknown,349


**Qué hacemos:** agrupamos la actividad, escrita en texto libre, en seis categorías: surfing, swimming, fishing, diving, other y unknown.

**Por qué:** sin esta columna no podemos separar los ataques a surfistas del resto, y todo el análisis pasaría a ser sobre ataques de tiburón en general, que no es lo que nos pidieron. La columna tiene cientos de valores distintos, y el surf aparece como "Surfing", "Body surfing", "Boogie boarding", "Paddle boarding", "Kite surfing"...

**La función por dentro:** `apply(funcion)` aplica nuestra función a cada valor de la columna, uno a uno, y devuelve una Serie con los resultados. Dentro de la función, `in` pregunta si una palabra está contenida en el texto, no si el texto es igual a esa palabra: por eso "Boogie boarding" encuentra "boogie" y entra en el grupo. El orden de los `if` importa: el primero que se cumple gana.

**Por qué una columna nueva y no sustituir `activity`:** guardamos el texto original para poder comprobar si la clasificación tiene sentido, con `new_df_shark_attacks[['activity', 'activity_group']].sample(20)`.

**Qué leemos:** 1593 ataques a surfistas, 1151 a nadadores, 1022 a pescadores y 529 a buceadores. El surf es la actividad con más ataques del archivo, y por eso el caso de negocio tiene sentido.

NOTA: es una aproximación hecha con palabras clave. Algún caso quedará en la categoría equivocada, y los vacíos quedan como `unknown`.

- limpieza para sex

In [58]:
new_df_shark_attacks['sex'] = new_df_shark_attacks['sex'].astype(str).str.strip().str.upper()
valores_validos = ['M', 'F']
new_df_shark_attacks['sex'] = new_df_shark_attacks['sex'].apply(lambda x: x if x in valores_validos else 'Unknown')

dejamos la columna con tres valores: `M`, `F` y `Unknown`.

**La función por dentro:** `apply(lambda x: x if x in valores_validos else 'Unknown')` recorre la columna valor a valor. Un `lambda` es una función escrita en una línea, sin nombre: recibe `x` y devuelve `x` si está en la lista, y `'Unknown'` si no.

Nota: entra en el perfil de la víctima, junto con la edad. No responde a ninguna de las cuatro preguntas principales, así que es un extra, no una conclusión.

- limpieza para age

In [59]:
new_df_shark_attacks['age'] = (new_df_shark_attacks['age']
                               .astype(str).str.extract(r'(\d+)')[0])

new_df_shark_attacks['age'] = pd.to_numeric(
    new_df_shark_attacks['age'], errors='coerce')

new_df_shark_attacks.loc[new_df_shark_attacks['age'] > 100, 'age'] = None
new_df_shark_attacks['age'] = new_df_shark_attacks['age'].astype('Int64')

new_df_shark_attacks['age'].describe()

,age
count,3563.0
mean,28.852372
std,14.883738
min,1.0
25%,17.0
50%,25.0
75%,38.0
max,87.0


sacamos el primer número de cada edad, la convertimos en entero y descartamos las imposibles.

**Por qué era texto:** la columna mezcla números con palabras: "30s", "teen", "18 or 20", "mid-20s". Con un solo valor así, pandas guarda toda la columna como texto y ninguna media se puede calcular.

**La función por dentro:** `str.extract(r'(\d+)')` busca el patrón en cada valor y devuelve lo que esté dentro del paréntesis, que es lo que se llama un grupo de captura. `\d+` significa "uno o más dígitos". Así "30s" da 30, "18 or 20" da 18 y "teen" queda vacío. Devuelve una tabla de una columna, y por eso lleva `[0]` al final para quedarnos con la columna.

**Por qué descartamos las mayores de 100:** son errores de registro. Una edad de 500 no es un dato raro, es un dato falso, y arrastraría la media hacia arriba sin que se note en el gráfico.

**Por qué `Int64` y no `int`:** siguen faltando muchas edades (en el archivo original, 2994 de 7125, el 42%). El `int` normal no admite vacíos y daría error.

**Qué leemos en `describe()`:** 3563 edades registradas de 5199 filas, así que falta el 31%. Media de 28,9 años y mediana de 25. El 25% es 17: al menos una de cada cuatro víctimas con edad tenía 17 años o menos.

NOTA: la edad no responde a ninguna de las cuatro preguntas principales, pero cuenta como técnica de limpieza y deja la columna lista para el perfil de la víctima.

- limpieza para injury

In [60]:
new_df_shark_attacks['injury'].value_counts(dropna=False)

,count
injury,
FATAL,382
Foot bitten,83
Survived,70
No injury,67
Leg bitten,58
...,...
Laceration to shin of left leg,1
Injury to forearm from shark's spine PROVOKED INCIDENT,1
Cause of death was drowning; his remains were scavenged by sharks,1


In [61]:
new_df_shark_attacks['injury_clean'] = new_df_shark_attacks['injury'].astype(str).str.strip().str.lower()
new_df_shark_attacks['injury_clean'].value_counts()

,count
injury_clean,
fatal,390
foot bitten,83
no injury,77
survived,70
leg bitten,61
...,...
"3 puncture wounds to big toe of left foot, and laceration on arch of foot",1
bitten on upper right thigh & right ankle,1
arm & torso bitten,1


In [62]:
mordeduras = new_df_shark_attacks['injury_clean'].str.contains('bite|bitten', regex=True).sum()
amputaciones = new_df_shark_attacks['injury_clean'].str.contains('amputat', regex=True).sum()
miembros_inferiores = new_df_shark_attacks['injury_clean'].str.contains('leg|foot|thigh|calf', regex=True).sum()
print(f"Mordeduras: {mordeduras}, Amputaciones: {amputaciones}, Miembros inferiores: {miembros_inferiores}")

Mordeduras: 1472, Amputaciones: 56, Miembros inferiores: 2143


In [63]:
import numpy as np

condiciones = [
    new_df_shark_attacks['injury_clean'].str.contains('fatal|death', regex=True, na=False),
    new_df_shark_attacks['injury_clean'].str.contains('amputat', regex=True, na=False),
    new_df_shark_attacks['injury_clean'].str.contains('no injury|uninjured', regex=True, na=False),
    new_df_shark_attacks['injury_clean'].str.contains('leg|foot|thigh|calf', regex=True, na=False),
    new_df_shark_attacks['injury_clean'].str.contains('arm|hand|shoulder|finger', regex=True, na=False)
]

categorias = ['fatal', 'amputation', 'no injury', 'lower limb injury', 'upper limb injury']

new_df_shark_attacks['injury_grouped'] = np.select(condiciones, categorias, default='other')

In [64]:
new_df_shark_attacks['injury_grouped'].value_counts()

,count
injury_grouped,
lower limb injury,1948
other,979
upper limb injury,783
no injury,718
fatal,717
amputation,54


normalizamos el texto, contamos algunos patrones y creamos una columna nueva, `injury_grouped`, con la gravedad de la herida.

**Por qué esta columna no se puede usar tal cual:** tiene miles de valores distintos (la salida de `value_counts()` lo dice en `Length`). Cada persona describió la herida a su manera, y el valor más repetido, "FATAL", no describe ninguna herida.

**La función por dentro, `str.contains`:** devuelve `True` donde el texto contiene el patrón. Con `regex=True`, la barra vertical significa "o": `'bite|bitten'` encuentra las dos palabras de una vez. Y como `True` vale 1, un `.sum()` detrás cuenta los casos.

**La función por dentro, `np.select`:** recibe una lista de condiciones y una lista de etiquetas, y a cada fila le pone la etiqueta de la **primera** condición que cumple. Por eso el orden importa: "fatal" va primero, para que un ataque mortal con la pierna herida cuente como mortal y no como herida en la pierna. Lo que no cumple ninguna recibe el `default`, "other".

**Qué leemos:** 1948 heridas en piernas y pies, 783 en brazos y manos, 718 sin herida, 717 mortales y 54 amputaciones. Otras 979 quedan en `other`: descripciones que no encajan en ninguna palabra clave.

NOTA: comparamos los mortales según las dos columnas, `fatality == 'Y'` y `injury_clean` conteniendo "fatal", en la celda de verificación, más abajo.

- limpieza para fatality

In [65]:
new_df_shark_attacks['fatality'] = (new_df_shark_attacks['fatality']
                                    .astype(str).str.strip().str.upper()
                                    .map({'Y': 'Y', 'N': 'N'}))

In [66]:
new_df_shark_attacks['fatality'].value_counts(dropna=False)

,count
fatality,
N,4050
Y,713
NaN,436


normalizamos el texto y dejamos únicamente `Y` y `N`. Todo lo demás pasa a
vacío.

**La función por dentro:** `map({'Y': 'Y', 'N': 'N'})` parece que no hace nada, porque cambia
cada valor por sí mismo. Pero `map` **vacía todo lo que no esté en el diccionario**, así que
`UNKNOWN`, `M`, `2017` o `y ` desaparecen de golpe. Es un filtro disfrazado de traducción, y es
justo el caso donde `map` es mejor que `replace`: aquí sí queremos cerrar la columna.

**Por qué no rellenamos los vacíos con `N`:** un `UNKNOWN` no significa "no murió". Rellenarlo
sería inventar datos, y como la mortalidad se calcula como muertos entre casos, cada `N`
inventado bajaría el porcentaje.

NOTA: la cuenta de `Y`, la de `N` y la de vacíos. Los vacíos se quedan fuera de cualquier porcentaje, y hay que decir cuántos son.

- limpieza para time

In [67]:
def sacar_hora(texto):
    # «1015hrs», «0945hrs», «1635 hrs»: las dos primeras cifras son la hora
    encontrado = re.search(r'\b(\d{1,2})(\d{2})\s*hrs?\b', texto)
    if encontrado:
        return int(encontrado.group(1))
    # «14h00», «11h30», «14h»: la hora va delante de la h
    encontrado = re.search(r'\b(\d{1,2})\s*h', texto)
    if encontrado:
        return int(encontrado.group(1))
    # «10:30»
    encontrado = re.search(r'\b(\d{1,2}):\d{2}', texto)
    if encontrado:
        return int(encontrado.group(1))
    return None

def sacar_franja(valor):
    texto = str(valor).lower()
    hora = sacar_hora(texto)
    if hora is not None:
        if hora <= 5:
            return 'night'
        elif hora <= 11:
            return 'morning'
        elif hora <= 17:
            return 'afternoon'
        elif hora <= 23:
            return 'evening'
        else:
            return 'unknown'
    if 'morning' in texto or 'dawn' in texto:
        return 'morning'
    elif 'afternoon' in texto or 'midday' in texto or 'noon' in texto:
        return 'afternoon'
    elif 'evening' in texto or 'dusk' in texto or 'sunset' in texto:
        return 'evening'
    elif 'night' in texto:
        return 'night'
    else:
        return 'unknown'

# prueba con ejemplos reales del archivo antes de fiarnos
for ejemplo in ['1015hrs', '0945hrs', '1100hrs', '1635 hrs', '14h00', '11h30', 'Late afternoon', 'Dusk']:
    print(ejemplo, '->', sacar_franja(ejemplo))

new_df_shark_attacks['time_group'] = new_df_shark_attacks['time'].apply(sacar_franja)
new_df_shark_attacks['time_group'].value_counts(dropna=False)

1015hrs -> morning
0945hrs -> morning
1100hrs -> morning
1635 hrs -> afternoon
14h00 -> afternoon
11h30 -> morning
Late afternoon -> afternoon
Dusk -> evening


,count
time_group,
unknown,2129
afternoon,1666
morning,1038
evening,287
night,79


convertimos la columna en cuatro franjas: night, morning, afternoon y evening.

**El problema:** la columna mezcla horas escritas de varias formas ("14h00", "11h30", "1015hrs", "1635 hrs") con palabras ("Afternoon", "late afternoon", "Dusk"). No es ni número ni categoría, es las dos cosas a la vez.

**El fallo que tenía la versión anterior:** buscaba "uno o dos dígitos seguidos de h". En "1015hrs" lo primero que cumple eso es "15h", así que las 10:15 salían como tarde; "1100hrs" daba "00h" y salía como noche; "0945hrs" daba "45h" y acababa en unknown. El formato de cuatro cifras es el de los registros recientes, justo los que más pesan en el análisis.

**La función por dentro:** `sacar_hora` prueba tres formatos, del más concreto al más general. Primero cuatro cifras seguidas de "hrs": `(\d{1,2})(\d{2})` separa la hora de los minutos, y nos quedamos con la hora. Después una o dos cifras seguidas de h, como en "14h00". Y por último "10:30". `\b` marca el borde de una palabra, para no empezar a leer a mitad de un número. Si no hay ninguna hora, `sacar_franja` busca palabras clave.

**Por qué la franja y no la hora exacta:** con la hora exacta tendríamos 24 barras y casi ningún caso en cada una. Cuatro franjas se leen de un vistazo, y es lo que una escuela necesita para decidir a qué horas meter a los alumnos en el agua.

**Qué leemos:** primero los ocho ejemplos de prueba, que tienen que salir bien antes de fiarse del resto: 1015hrs y 0945hrs, morning; 1635 hrs, afternoon. Después, 1666 ataques por la tarde, 1038 por la mañana, 287 entre las 18 y las 23 h, 79 de madrugada y 2129 sin hora.

NOTA: `time` está vacía en el 49,5% del archivo original. Cualquier conclusión sobre la hora se basa en los casos que sí la tienen, y eso va en la diapositiva, no en una nota al pie.

In [68]:
new_df_shark_attacks = new_df_shark_attacks.reset_index(drop=True)

print('Filas:', len(new_df_shark_attacks))
print('Columnas:', len(new_df_shark_attacks.columns))
new_df_shark_attacks.info()

Filas: 5199
Columnas: 19
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5199 entries, 0 to 5198
Data columns (total 19 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   month           4966 non-null   Int64 
 1   year            5199 non-null   Int64 
 2   type            5199 non-null   object
 3   country         5199 non-null   object
 4   state           5199 non-null   object
 5   location        5199 non-null   object
 6   activity        4850 non-null   object
 7   sex             5199 non-null   object
 8   age             3563 non-null   Int64 
 9   injury          5178 non-null   object
 10  fatality        4763 non-null   object
 11  time            3165 non-null   object
 12  season          5199 non-null   object
 13  bloque_año      5199 non-null   object
 14  state_group     5199 non-null   object
 15  activity_group  5199 non-null   object
 16  injury_clean    5199 non-null   object
 17  injury_grouped  5199 non-nu

volvemos a numerar las filas y hacemos el recuento final.

**La función por dentro:** al borrar filas, el índice queda con huecos (0, 1, 5, 9...). `reset_index()` vuelve a numerar desde 0, y `drop=True` tira el índice viejo. Sin `drop=True`, pandas lo guarda como una columna nueva llamada `index` que no sirve para nada y ensucia la tabla.

**Por qué al final y no antes:** cada paso que borra filas vuelve a abrir huecos. Numerar en medio de la limpieza obliga a repetirlo.

**Qué leemos:** el `info()` final es la mejor foto del trabajo hecho. Se compara con el `info()` del principio y se ve todo de una vez: las columnas que quedaron, los tipos que cambiaron (`year` ya entero, `age` ya `Int64`) y cuántos vacíos siguen en cada columna.

**Resultado:** de 7125 filas y 23 columnas a 5199 filas y 19 columnas. Eso resume toda la limpieza en una línea.

# PARTE 3. ANÁLISIS

De la tabla limpia nos quedamos con las filas que responden a nuestro caso.

**Los dos filtros:** `type == 'unprovoked'`, porque un ataque provocado, como el de un pescador que sube un tiburón al barco, no dice nada del riesgo de surfear. Y `activity_group == 'surfing'`, porque el caso de negocio es una escuela de surf, no el riesgo general de bañarse.

**Por qué filtramos aquí y no durante la limpieza:** la tabla limpia se queda entera. Así podemos comparar el surf con las demás actividades, que es lo que responde a la primera pregunta, y contestar a otras cosas sin volver a limpiar nada.

NOTA: si al final quedan pocos casos, todos los porcentajes que salgan de aquí se calculan sobre una base pequeña y hay que presentarlos siempre con el número de casos al lado.

In [69]:
# todo el análisis de recomendación vive en el bloque más reciente
reciente = new_df_shark_attacks[new_df_shark_attacks['bloque_año'] == BLOQUE_RECIENTE]

surf_mundo = reciente[(reciente['type'] == 'unprovoked') &
                      (reciente['activity_group'] == 'surfing')]

top_10_estados = (surf_mundo['state']
                  .value_counts()
                  .drop('unknown', errors='ignore')
                  .head(10).index.tolist())

df_top10 = reciente[(reciente['type'] == 'unprovoked') &
                    (reciente['state'].isin(top_10_estados))]
df_surf = df_top10[df_top10['activity_group'] == 'surfing']

print('Bloque:', BLOQUE_RECIENTE)
print('Registros en el bloque:', len(reciente))
print('Top 10 estados (por ataques a surfistas):', top_10_estados)
print('No provocados en el top 10:', len(df_top10))
print('Surfistas en el top 10:', len(df_surf))
print('Surfistas en todo el mundo:', len(surf_mundo))

Bloque: 2001-2025
Registros en el bloque: 2714
Top 10 estados (por ataques a surfistas): ['florida', 'new south wales', 'hawaii', 'california', 'western australia', 'eastern cape province', 'western cape province', 'south carolina', 'north carolina', 'south australia']
No provocados en el top 10: 1350
Surfistas en el top 10: 755
Surfistas en todo el mundo: 967


### Las tablas del análisis, todas en el bloque más reciente

**Qué hacemos:** definimos de una vez las tablas con las que vamos a trabajar, antes de las preguntas, para que todas usen el bloque 2001-2025.

**Por qué el bloque reciente y no el archivo entero:** una escuela que abre hoy quiere conocer el riesgo de hoy. Los bloques anteriores sirven para contar cómo ha cambiado, no para recomendar.

**Por qué esta celda va antes de las preguntas:** antes estaba al final, y las preguntas 1 a 4 se habían calculado con todos los años. Una tabla que se redefine después de usarse no cambia lo que ya se calculó con ella.

**Por qué el top 10 se calcula con los ataques a surfistas:** para decidir dónde no abrir una escuela de surf, el ranking que importa es el de ataques a surfistas, no el de ataques en general. Quitamos `unknown`, que no es un estado sino los ataques sin estado registrado.

**La función por dentro:** `reciente['type'] == 'unprovoked'` devuelve una columna de `True` y `False`, y `tabla[condición]` se queda con las filas `True`. El `&` es el "y" de pandas y cada condición va entre paréntesis, porque si no Python evalúa el `&` antes que el `==` y da error.

**Las cuatro tablas, y para qué sirve cada una:**
- `reciente` es el bloque entero, con todas las actividades.
- `surf_mundo` son los ataques no provocados a surfistas en todo el mundo. Busca destinos seguros, y es la única que no recorta por estado: los sitios sin ataques están justamente fuera del top 10.
- `df_top10` compara el surf con las otras actividades, así que necesita las demás dentro.
- `df_surf` responde a cuándo y dónde, y solo tiene surfistas.

**Qué leemos:** 2714 registros en el bloque, 967 ataques no provocados a surfistas en el mundo y 755 en los diez estados con más casos. El top 10 lo encabezan Florida, Nueva Gales del Sur, Hawái y California.

In [70]:
tabla_actividad = (pd.crosstab(df_top10['activity_group'],
                               df_top10['injury_grouped'],
                               normalize='index') * 100).round(1)
tabla_actividad['casos'] = df_top10['activity_group'].value_counts()
tabla_actividad

injury_grouped,amputation,fatal,lower limb injury,no injury,other,upper limb injury,casos
activity_group,,,,,,,
diving,1.2,16.0,32.1,8.6,19.8,22.2,81
fishing,1.4,6.8,30.1,11.0,21.9,28.8,73
other,0.0,1.7,50.8,10.2,16.9,20.3,59
surfing,0.1,5.2,49.5,19.7,12.7,12.7,755
swimming,1.2,6.3,62.8,0.6,13.9,15.1,331
unknown,0.0,2.0,51.0,0.0,29.4,17.6,51


### Pregunta 1: el surf frente a otras actividades

Cruzamos actividad con gravedad de la herida y lo pasamos a porcentaje dentro de cada actividad.

**La función por dentro:** `pd.crosstab(filas, columnas)` cuenta cuántas veces aparece cada combinación y devuelve una tabla de actividades por gravedades. Con `normalize='index'` divide cada fila entre su total, así que cada fila suma 1; multiplicado por 100, son porcentajes. La columna `casos` pone al lado sobre cuántos ataques se calcula cada fila.

**Por qué `crosstab` y no `groupby(...).apply(...)`:** la versión anterior hacía lo mismo en dos pasos y devolvía el nombre de la actividad repetido en dos niveles del índice. `crosstab` lo deja en una tabla que se lee de un vistazo y que se puede llevar a una diapositiva.

**Por qué porcentaje por actividad y no sobre el total:** hay muchos más ataques a unas actividades que a otras, así que en números absolutos la comparación mediría cuánta gente hace cada cosa. Dividiendo dentro de cada actividad comparamos la gravedad de los ataques, que es lo que queremos saber.

**Qué leemos:** en el surf, el 5,2% de los ataques son mortales, frente al 6,3% en natación, el 6,8% en pesca y el 16,0% en buceo. Y el 19,7% de los ataques a surfistas acaban sin herida, frente al 0,6% en natación. Ojo con las bases: pesca y buceo tienen 73 y 81 casos, así que sus porcentajes se mueven mucho con pocos ataques.

**NOTA:** el surf tiene menos porcentaje de casos mortales que la natación. Eso es lo que muestran los datos y se puede afirmar. Lo que no podemos afirmar es el porqué: la tabla como escudo es una hipótesis razonable, no algo que estos datos demuestren. Y cada porcentaje necesita su número de casos: un 50% sobre dos ataques no dice nada.

In [71]:
orden_franjas = ['morning', 'afternoon', 'evening', 'night']

con_hora = df_surf[df_surf['time_group'] != 'unknown']
pct_franja = (con_hora['time_group']
              .value_counts(normalize=True)
              .reindex(orden_franjas, fill_value=0) * 100).round(1)

print('Surfistas en el top 10:', len(df_surf))
print('Con hora registrada:', len(con_hora))
pct_franja

Surfistas en el top 10: 755
Con hora registrada: 667


,proportion
time_group,
morning,42.4
afternoon,51.1
evening,6.3
night,0.1


### Pregunta 2: a qué hora ocurren los ataques a surfistas

Sobre los surfistas con hora registrada, calculamos el porcentaje de ataques en cada franja.

**Las funciones por dentro:** `value_counts(normalize=True)` cuenta cada franja y la divide entre el total, así que devuelve proporciones que suman 1. `reindex(orden_franjas)` pone las franjas en el orden del día y no de mayor a menor, que es como se lee un horario.

**Por qué quitamos `unknown` de la base:** si se queda dentro, cada porcentaje baja según cuántas horas falten, y la franja más alta podría acabar siendo "no sabemos". Calculamos sobre los que tienen hora e imprimimos cuántos son, que es lo que hay que decir al presentar.

**Qué leemos:** de 755 surfistas, 667 tienen hora registrada. De ellos, el 51,1% de los ataques ocurren por la tarde (12 a 17 h) y el 42,4% por la mañana (6 a 11 h). Entre las 18 y las 23 h, el 6,3%, y de madrugada casi ninguno. La diferencia entre tarde y mañana es de nueve puntos: no basta para decir que la mañana sea segura.

**NOTA:** que haya menos ataques en una franja no significa que sea más segura: puede significar que hay menos gente en el agua a esa hora. Los datos cuentan ataques, no surfistas. La frase honesta es del tipo "la mayoría de los ataques registrados ocurren en tal franja", nunca "esa franja es más peligrosa". Aun así, una escuela decide a qué horas mete a los alumnos en el agua, y para eso basta saber cuándo se concentran los ataques registrados.

In [72]:
orden_estaciones = ['primavera', 'verano', 'otoño', 'invierno']

con_mes = df_surf[df_surf['season'] != 'unknown']
pct_estacion = (pd.crosstab(con_mes['state'], con_mes['season'], normalize='index')
                .reindex(columns=orden_estaciones, fill_value=0) * 100).round(1)
pct_estacion['casos'] = con_mes['state'].value_counts()

pct_estacion.sort_values('casos', ascending=False)

season,primavera,verano,otoño,invierno,casos
state,,,,,
florida,28.9,28.3,38.1,4.8,315
new south wales,26.5,32.5,21.4,19.7,117
hawaii,24.4,15.4,33.3,26.9,78
california,12.9,28.6,45.7,12.9,70
western australia,34.0,14.9,23.4,27.7,47
eastern cape province,14.3,28.6,32.1,25.0,28
western cape province,28.6,7.1,28.6,35.7,28
north carolina,4.5,81.8,9.1,4.5,22
south carolina,9.1,72.7,18.2,0.0,22


### Pregunta 3: en qué estación se concentran los ataques, estado por estado

 para cada estado, el porcentaje de ataques a surfistas en cada estación, sobre los que tienen mes.

**Por qué por estado y no en total:** un total mezclaría Florida con Nueva Gales del Sur, que están en hemisferios distintos. El verano de uno es el invierno del otro, y sumados se anulan: el gráfico saldría plano y la conclusión sería que la estación no importa.

**La función por dentro:** `pd.crosstab` cuenta cada combinación de estado y estación, y con `normalize='index'` divide cada fila entre su total, así que cada estado suma 100. `reindex(columns=...)` pone las estaciones en el orden del año, y la columna `casos` dice sobre cuántos ataques se calcula cada fila.

**Por qué porcentaje y no recuento:** los números en bruto se comparan mal entre estados grandes y pequeños. Con el porcentaje dentro de cada estado se ve la temporada de cada uno.

**Qué leemos:** no hay una temporada única. En Florida el 38,1% de los ataques son en otoño, y en California el 45,7%. En las dos Carolinas se concentran en verano: 81,8% en la del Norte y 72,7% en la del Sur, aunque sobre solo 22 casos cada una. En Hawái y en Australia Occidental se reparten por todo el año.

NOTA: la columna `season` tiene el hemisferio corregido, así que "verano" quiere decir verano local en cada sitio. De aquí sale el entregable más accionable: un calendario de riesgo estacional por destino, que permite reforzar la vigilancia en meses concretos sin cerrar el resto del año.

In [73]:
df_surf.groupby(['state', 'sex'])['age'].agg(['count', 'mean', 'min', 'max']).round(1)

count  mean   min   max
state                 sex                             
california            F            6  33.7    15    54
                      M           41  33.6    16    61
                      Unknown      1  38.0    38    38
eastern cape province F            1  13.0    13    13
                      M           23  26.5    15    46
florida               F           33  23.8     8    64
                      M          249  26.2     6    86
                      Unknown      3  16.7    15    19
hawaii                F            7  29.7    13    54
                      M           48  38.2    10    71
                      Unknown      0  <NA>  <NA>  <NA>
new south wales       F            6  28.8    13    52
                      M           89  34.3    10    71
                      Unknown      0  <NA>  <NA>  <NA>
north carolina        F            4  14.2     6    28
                      M           15  23.5     9    54
                      Unknown      0  <NA>  <NA>  <NA>
south australia       M           18  34.2    15    64
south carolina        F           10  18.9     6    68
                      M           12  18.8     9    60
western australia     F            1  17.0    17    17
                      M           35  32.8    11    75
                      Unknown      0  <NA>  <NA>  <NA>
western cape province M           24  30.0    14    55

### Pregunta 4: perfil de edad y sexo de los surfistas atacados

agrupamos por estado y sexo y calculamos el número de casos, la media, el mínimo y el máximo de la edad.

**La función por dentro:** `agg(['count', 'mean', 'min', 'max'])` aplica varias funciones de una vez a la misma columna y devuelve una tabla con una columna por cada una.

**Por qué `count` y no `size` aquí:** `size` contaría las filas del grupo, incluidas las que no tienen edad. `count` cuenta solo las edades registradas, que son las que sostienen la media. Las dos juntas miden además la calidad del dato en cada estado.

.

**NOTA:** si en un estado las víctimas son más jóvenes, eso describe a quién surfea allí tanto como al riesgo. Un sitio con muchos surfistas jóvenes tendrá víctimas jóvenes sin ser más peligroso para ellos. Es un dato de perfil, no de riesgo.

OUTRA NOTA: son los valores extremos, y un solo caso raro los mueve entero. Se enseñan como contexto, nunca como conclusión.

**Qué leemos:**

- **Víctimas más jóvenes** en Carolina del Sur (18,8 años de media en los hombres), Carolina del Norte (23,5) y Florida (26,2), con víctimas de hasta 6 años en los tres. Para una escuela con grupos infantiles, son los sitios donde más hay que reforzar la vigilancia.
- **Víctimas adultas** en Hawái (38,2), Nueva Gales del Sur (34,3), Australia Meridional (34,2), California (33,6) y Australia Occidental (32,8): la media pasa de 30 años.
- En Australia Meridional y en Cabo Occidental solo hay ataques registrados a hombres.

NOTA: más joven no significa más peligroso para los jóvenes. Si en un estado surfea gente joven, sus víctimas serán jóvenes aunque el riesgo sea el mismo. Es un dato de perfil, no de riesgo.

In [74]:
print('fatality == Y:', (new_df_shark_attacks['fatality'] == 'Y').sum())
print('injury dice fatal:',
      new_df_shark_attacks['injury_clean'].str.contains('fatal', na=False).sum())
print('fatality vacía:', new_df_shark_attacks['fatality'].isna().sum())

discrepancia = new_df_shark_attacks[
    (new_df_shark_attacks['fatality'] != 'Y') &
    (new_df_shark_attacks['injury_clean'].str.contains('fatal', na=False))]

print('Casos donde injury dice fatal pero fatality no:', len(discrepancia))
discrepancia[['country', 'year', 'fatality', 'injury']].head(10)

fatality == Y: 713
injury dice fatal: 674
fatality vacía: 436
Casos donde injury dice fatal pero fatality no: 30


,country,year,fatality,injury
219,NEW CALEDONIA,2023,NaN,"FATAL, Multiple injuries"
803,SOUTH AFRICA,2017,NaN,"FATAL, but death was probably due to drowning"
811,SOUTH AFRICA,2017,NaN,"FATAL, but death was probably due to drowning"
886,SOUTH AFRICA,2017,NaN,"Fatal, coroner unable to determine if the dive..."
1169,ITALY,2015,NaN,"FATAL, but shark involvement prior to death un..."
1605,AUSTRALIA,2011,NaN,FATAL
2227,SOUTH AFRICA,2006,NaN,"FATAL, but shark involvement prior to death un..."
2295,SEYCHELLES,2005,NaN,"FATAL, shark involvement prior to death is unc..."
2611,NEW CALEDONIA,2002,NaN,"PRESUMED FATAL, body not recovered"
2679,MADAGASCAR,2001,NaN,Fatal or drowned & remains scavenged by shark


contamos los ataques mortales por dos caminos independientes y comparamos.

**Por qué esta celda existe:** `fatality` es una columna de una sola letra que alguien rellenó a la mano, y de ella sale la tasa de mortalidad, que es uno de los números que vamos a presentar. Antes
de fiarnos, la contrastamos con otra fuente del mismo archivo: el texto de `injury`.


NOTA: `injury` es texto libre, y
alguna descripción dirá "fatal" sin serlo, o al revés. Lo que buscamos es que se parezcan. Si
`fatality` dijera 1500 y el texto dijera 400, la columna no sería de fiar y habría que
reconstruirla desde `injury`.

 los casos donde el texto dice mortal y la columna no. Basta mirar
diez para saber si es ruido del texto libre o un fallo real del registro.

 si cuadra, `injury` ya cumplió su función y se puede quitar al final. Y
esta celda se queda como diapositiva de verificación, que es de las que mejor impresión dan.

**Qué leemos:** `fatality` marca 713 mortales y el texto de `injury` dice "fatal" en 674. Se parecen, así que `fatality` es de fiar y la mortalidad se puede presentar. En los 30 casos en que el texto dice mortal y la columna no, `fatality` está vacío y el propio texto duda: "probably drowning", "shark involvement prior to death unconfirmed". No es un fallo del registro: son casos en los que nadie sabe si el tiburón causó la muerte, y dejarlos fuera del porcentaje es lo correcto.

In [75]:
evolucion = (new_df_shark_attacks
             .groupby('bloque_año')
             .agg(ataques=('country', 'size'),
                  con_fatality=('fatality', 'count'),
                  mortales=('fatality', lambda s: (s == 'Y').sum()),
                  paises=('country', 'nunique')))

evolucion['mortalidad'] = (evolucion['mortales'] / evolucion['con_fatality'] * 100).round(1)
evolucion

,ataques,con_fatality,mortales,paises,mortalidad
bloque_año,,,,,
1951-1975,1262,1127,284,102,25.2
1976-2000,1223,1108,180,53,16.2
2001-2025,2714,2528,249,105,9.8


Para cada bloque de 25 años: cuántos ataques hay, cuántos tienen desenlace registrado, cuántos fueron mortales, en cuántos países ocurrieron y qué porcentaje fueron mortales.

**La métrica que sí se puede comparar entre bloques: la mortalidad.** Es un porcentaje dentro de cada bloque, así que no depende de cuántos casos se registraran. Si baja del bloque antiguo al reciente, eso dice algo real sobre rescate, medicina y tiempo de respuesta, y no sobre la calidad del archivo.

**La métrica que NO se puede comparar: el número de ataques.** Sube hacia el presente porque hoy se documenta casi todo y hace setenta años no. Se enseña, pero con la frase al lado: esto mide el registro tanto como el suceso.

**Por qué la mortalidad se divide entre `con_fatality` y no entre `ataques`:** un ataque sin desenlace registrado no es un ataque en el que nadie murió. Si lo dejáramos en el denominador, contaría como superviviente y bajaría el porcentaje, que es justo lo que decidimos no hacer al limpiar `fatality`. La versión anterior, `(s == 'Y').mean()`, cometía ese error sin que se viera.

**Las funciones por dentro:** `groupby('bloque_año')` agrupa por la columna de bloques; como los nombres empiezan por el año, salen en orden cronológico. `size` cuenta filas, `count` cuenta valores rellenos y `nunique` cuenta valores distintos. El `lambda` suma los `Y`, porque `True` vale 1.

**Qué leemos:** la mortalidad pasó del 25,2% en 1951-1975 al 16,2% en 1976-2000 y al 9,8% en 2001-2025: menos de la mitad. Es un dato que tranquiliza al cliente y que ningún grupo que solo mire los últimos años va a tener.

In [76]:
destinos_surf = [
    # los tres grandes del surf, que son también los de más ataques
    'USA', 'AUSTRALIA', 'SOUTH AFRICA',
    # Europa y norte de África
    'PORTUGAL', 'SPAIN', 'FRANCE', 'MOROCCO',
    # América Central y del Sur
    'MEXICO', 'COSTA RICA', 'NICARAGUA', 'EL SALVADOR', 'PANAMA',
    'PERU', 'ECUADOR', 'CHILE', 'BRAZIL',
    # Asia y Pacífico
    'INDONESIA', 'PHILIPPINES', 'SRI LANKA', 'MALDIVES', 'FIJI', 'NEW ZEALAND',
]

paises_en_datos = set(new_df_shark_attacks['country'].unique())

faltan = [p for p in destinos_surf if p not in paises_en_datos]
print('Nombres que NO existen en la columna country:', faltan)

for palabra in ['MALDIV', 'ISLAND', 'RICA', 'SALVADOR', 'PHILIPP', 'ZEALAND']:
    print(palabra, '->', sorted(p for p in paises_en_datos if palabra in p))

Nombres que NO existen en la columna country: ['PERU']
MALDIV -> ['MALDIVES']
ISLAND -> ['ADMIRALTY ISLANDS', 'ANDAMAN / NICOBAR ISLANDAS', 'BRITISH VIRGIN ISLANDS', 'CANARY ISLANDS', 'CAYMAN ISLANDS', 'JOHNSTON ISLAND', 'MARSHALL ISLANDS', 'SOLOMON ISLANDS']
RICA -> ['AMERICAN SAMOA', 'COSTA RICA', 'SOUTH AFRICA']
SALVADOR -> ['EL SALVADOR']
PHILIPP -> ['PHILIPPINES']
ZEALAND -> ['NEW ZEALAND']


escribimos la lista de destinos y verificamos que cada nombre existe de verdad en
la columna `country`.


**La función por dentro:** `set()` sobre los valores únicos permite preguntar "¿está este nombre
aquí?" de forma instantánea. La comprensión de lista recorre nuestros destinos y se queda con los
que no aparecen. El bucle de abajo busca fragmentos, para encontrar cómo se llama de verdad cada
uno de los que faltan.


**El criterio de la lista, que hay que poder defender:** destinos de surf con escuelas ya operando,
temporada larga y que una escuela podría elegir de verdad. Incluimos a propósito Estados Unidos,
Australia y Sudáfrica aunque sean los peores: sin ellos la comparación no tiene escala y nadie sabe
si dos ataques es mucho o poco. Veintidós destinos: menos de quince parece elegido a dedo para dar
el resultado que queríamos, y más de veinticinco no cabe en un gráfico legible.

**Qué leemos en la comprobación:** todos los nombres existen salvo `PERU`. Perú no tiene ningún ataque registrado en la tabla filtrada, así que no aparece en `country`: no es una errata, es justo el caso para el que usamos `reindex` con `fill_value=0`. La búsqueda por fragmentos destapó además dos países partidos en dos nombres, `MALDIVES` y `MALDIVE ISLANDS`, `REUNION` y `REUNION ISLAND`, que unificamos en la limpieza de `country`.

In [77]:
ataques_por_pais = surf_mundo.groupby('country').size()

destinos = (ataques_por_pais
            .reindex(destinos_surf, fill_value=0)
            .sort_values())

print(f'Ataques a surfistas por destino, {BLOQUE_RECIENTE}')
print(destinos)
print()
print('Sin ningún ataque registrado:', list(destinos[destinos == 0].index))
print()
print('Ataques al año, en los destinos con alguno:')
print((destinos[destinos > 0] / 25).round(2))

Ataques a surfistas por destino, 2001-2025
country
PORTUGAL          0
FRANCE            0
MOROCCO           0
NICARAGUA         0
PERU              0
CHILE             0
EL SALVADOR       0
PHILIPPINES       0
SRI LANKA         0
MALDIVES          0
PANAMA            1
SPAIN             1
FIJI              2
COSTA RICA        4
ECUADOR           4
INDONESIA         7
MEXICO            9
NEW ZEALAND      14
BRAZIL           17
SOUTH AFRICA     65
AUSTRALIA       228
USA             556
dtype: int64

Sin ningún ataque registrado: ['PORTUGAL', 'FRANCE', 'MOROCCO', 'NICARAGUA', 'PERU', 'CHILE', 'EL SALVADOR', 'PHILIPPINES', 'SRI LANKA', 'MALDIVES']

Ataques al año, en los destinos con alguno:
country
PANAMA           0.04
SPAIN            0.04
FIJI             0.08
COSTA RICA       0.16
ECUADOR          0.16
INDONESIA        0.28
MEXICO           0.36
NEW ZEALAND      0.56
BRAZIL           0.68
SOUTH AFRICA     2.60
AUSTRALIA        9.12
USA             22.24
dtype: float64


 el archivo solo contiene lugares **donde hubo
ataques**. Un destino sin ningún ataque no aparece con un cero, no aparece en absoluto. Si ordenáramos el ranking al revés buscando el sitio más seguro, ganaría un país con un solo ataque y sin olas, que no le sirve de nada a una escuela de surf. Entonces, no preguntar "cuál es el país más seguro del mundo", que estos datos no pueden
responder, sino "de estos destinos que nos interesan, cuáles registran menos ataques". Comparamos
dentro de una lista cerrada, elegida y justificada por nosotros.

**La función por dentro, `reindex`:** coloca las filas en el orden de la lista que le damos y, con
`fill_value=0`, **crea con un cero** las que no existían. Es la diferencia entre "no hay datos" y
"hay un cero", y sin ella los mejores destinos desaparecerían del resultado.

**Por qué `sort_values()` sin `ascending=False`:** en las otras preguntas buscábamos el peor y
ordenábamos de mayor a menor. Aquí buscamos el mejor, así que el orden natural pone arriba lo que
nos interesa.

**Por qué dividimos entre 25:** el periodo dura exactamente 25 años, así que el número por año sale de una división limpia.

**Por qué usamos `surf_mundo` y no `df_surf`:** los sitios seguros están justamente fuera del top
10. Buscarlos dentro del top 10 sería buscar al mejor entre los diez peores.

NOTA: un cero significa "sin ataques registrados", no
"sin riesgo". Puede significar también que allí se registra peor. Y la lista es nuestra: otra lista
daría otro ranking.

**Qué leemos:** 10 de los 22 destinos no registran ningún ataque a surfistas en 2001-2025: Portugal, Francia, Marruecos, Nicaragua, Perú, Chile, El Salvador, Filipinas, Sri Lanka y Maldivas. En el otro extremo están Estados Unidos (556), Australia (228) y Sudáfrica (65). España, con 1, sale a 0,04 ataques al año: una cifra que se entiende en una reunión mucho mejor que el total.

In [78]:
surf_todos = new_df_shark_attacks[(new_df_shark_attacks['type'] == 'unprovoked') &
                                  (new_df_shark_attacks['activity_group'] == 'surfing')]

destinos_todos = (surf_todos.groupby('country').size()
                  .reindex(destinos_surf, fill_value=0))

print(f'A cero en el bloque {BLOQUE_RECIENTE}:', list(destinos[destinos == 0].index))
print(f'A cero en todo {ANIO_MIN}-{ANIO_MAX}:', list(destinos_todos[destinos_todos == 0].index))

A cero en el bloque 2001-2025: ['PORTUGAL', 'FRANCE', 'MOROCCO', 'NICARAGUA', 'PERU', 'CHILE', 'EL SALVADOR', 'PHILIPPINES', 'SRI LANKA', 'MALDIVES']
A cero en todo 1951-2025: ['PORTUGAL', 'FRANCE', 'MOROCCO', 'NICARAGUA', 'PERU', 'CHILE', 'SRI LANKA', 'MALDIVES']


**Qué hacemos:** repetimos el recuento de destinos con los tres bloques juntos, de 1951 a 2025.

**Por qué:** un cero en 25 años puede ser suerte. Un cero en 75 años es mucho más difícil de explicar por casualidad. Si un destino sigue a cero con todos los años, la recomendación se sostiene mejor.

**Qué leemos:** 8 destinos siguen sin ningún ataque registrado a surfistas desde 1951: Portugal, Francia, Marruecos, Nicaragua, Perú, Chile, Sri Lanka y Maldivas. El Salvador y Filipinas salen de la lista: tienen algún ataque antes de 2001. Esos ocho son los candidatos más sólidos para abrir.

NOTA: sigue siendo "sin ataques registrados", no "sin riesgo". Setenta y cinco años sin registros también pueden ser setenta y cinco años registrando peor.

In [79]:
resumen = (surf_mundo.groupby('country').agg(
        ataques=('country', 'size'),
        playas=('location', lambda s: s[s != 'nan'].nunique()),
        con_edad=('age', 'count'),
        edad_media=('age', 'mean'),
        con_fatality=('fatality', 'count'),
        mortales=('fatality', lambda s: (s == 'Y').sum()))
    .sort_values('ataques', ascending=False)
    .head(15))

resumen['mortalidad'] = (resumen['mortales'] / resumen['con_fatality'] * 100).round(1)
resumen.round(1)

,ataques,playas,con_edad,edad_media,con_fatality,mortales,mortalidad
country,,,,,,,
USA,556,287,460,27.7,554,12,2.2
AUSTRALIA,228,184,177,33.2,227,25,11.0
SOUTH AFRICA,65,50,52,28.8,65,8,12.3
REUNION,24,18,24,29.5,24,9,37.5
BRAZIL,17,14,16,26.6,17,2,11.8
NEW ZEALAND,14,12,10,25.5,14,0,0.0
NEW CALEDONIA,9,8,8,32.0,9,5,55.6
MEXICO,9,9,8,37.1,9,4,44.4
INDONESIA,7,3,3,30.0,7,0,0.0


Juntamos en una sola tabla el número de ataques, en cuántas playas distintas, la
edad media, cuántos tienen edad registrada y el porcentaje de casos mortales.

**La función por dentro, `agg` con nombres:** `agg(nombre=('columna', 'función'))` calcula varias
métricas de golpe y deja cada una con el nombre que le damos. Es mucho más legible que un `agg`
con una lista, donde después hay que adivinar qué columna es cuál.

**Las cinco métricas, y por qué cada una:**
- `size` cuenta **filas**, o sea ataques. No usamos `count`, que contaría valores rellenos de una
  columna y perdería los ataques con campos vacíos.
- `nunique` en `location` dice en cuántas playas distintas ocurrieron. Es la métrica que separa
  "país peligroso" de "playa peligrosa": un país con muchos ataques concentrados en una sola playa
  puede ser perfectamente seguro a 50 km de allí.
- `count` en `age` son las edades registradas. Va al lado de la media a propósito.
- `mean` en `age` es la edad media, calculada solo sobre las que existen.
- La mortalidad se calcula con un `lambda`: `(s == 'Y')` da una columna de `True` y `False`, y su
  media es directamente la proporción de mortales, porque `True` vale 1. Multiplicada por 100, es
  el porcentaje.
  
NOTA: ningún porcentaje sin su número de casos. Un país con 2
ataques y 1 mortal da un 50% de mortalidad y parecería el lugar más peligroso del mundo; con la
columna de ataques al lado se ve enseguida que no se puede concluir nada.

**Dos cambios respecto a la versión anterior:** la mortalidad se divide entre los ataques con desenlace registrado (`con_fatality`) y no entre todos, porque un vacío no es un superviviente. Y `playas` ya no cuenta `'nan'` como si fuera una playa más.

**Qué leemos:** Estados Unidos concentra 556 ataques en 287 playas distintas, con una mortalidad del 2,2%: el riesgo está repartido por todo el país, y por eso hay que bajar al nivel de estado y de playa. Reunión tiene 24 ataques, 9 de ellos mortales (37,5%): con más de diez casos, es el destino con la mortalidad más alta de la tabla. Los países con dos o tres ataques y un 33% o un 50% de mortalidad son exactamente el caso de la NOTA: no se puede concluir nada.